{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import muon\n",
    "import scanpy as sc\n",
    "import zarr\n",
    "import seqdata as sd\n",
    "import xarray as xr\n",
    "from anndata.experimental.pytorch import AnnLoader\n",
    "import torch\n",
    "from sklearn.preprocessing import OneHotEncoder, LabelEncoder\n",
    "\n",
    "import tempfile\n",
    "import anndata\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from typing import Optional, Union\n",
    "\n",
    "\n",
    "def _dna_to_code(nt: str) -> int:\n",
    "    if nt == \"A\":\n",
    "        return 0\n",
    "    elif nt == \"C\":\n",
    "        return 1\n",
    "    elif nt == \"G\":\n",
    "        return 2\n",
    "    elif nt == \"T\":\n",
    "        return 3\n",
    "    else:\n",
    "        # scBasset does this\n",
    "        return np.random.randint(0, 3)\n",
    "    \n",
    "\n",
    "def add_dna_sequence(\n",
    "    adata: anndata.AnnData,\n",
    "    seq_len: int = 1344,\n",
    "    genome_name: str = \"hg38\",\n",
    "    genome_dir: Optional[Path] = None,\n",
    "    genome_provider: Optional[str] = None,\n",
    "    install_genome: bool = True,\n",
    "    chr_var_key: str = \"chr\",\n",
    "    start_var_key: str = \"start\",\n",
    "    end_var_key: str = \"end\",\n",
    "    sequence_varm_key: str = \"dna_sequence\",\n",
    "    code_varm_key: str = \"dna_code\",\n",
    ") -> None:\n",
    "    \"\"\"Add DNA sequence to AnnData object.\n",
    "\n",
    "    Uses genomepy under the hood to download the genome.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata\n",
    "        AnnData object with chromatin accessiblity data\n",
    "    seq_len\n",
    "        Length of DNA sequence to extract around peak center.\n",
    "        Defaults to value used in scBasset.\n",
    "    genome_name\n",
    "        Name of genome to use, installed with genomepy\n",
    "    genome_dir\n",
    "        Directory to install genome to, if not already installed\n",
    "    genome_provider\n",
    "        Provider of genome, passed to genomepy\n",
    "    install_genome\n",
    "        Install the genome with genomepy. If False, `genome_provider` is not used,\n",
    "        and a genome is loaded with `genomepy.Genome(genome_name, genomes_dir=genome_dir)`\n",
    "    chr_var_key\n",
    "        Key in `.var` for chromosome\n",
    "    start_var_key\n",
    "        Key in `.var` for start position\n",
    "    end_var_key\n",
    "        Key in `.var` for end position\n",
    "    sequence_varm_key\n",
    "        Key in `.varm` for added DNA sequence\n",
    "    code_varm_key\n",
    "        Key in `.varm` for added DNA sequence, encoded as integers\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "\n",
    "    Adds fields to `.varm`:\n",
    "        sequence_varm_key: DNA sequence\n",
    "        code_varm_key: DNA sequence, encoded as integers\n",
    "    \"\"\"\n",
    "    import genomepy\n",
    "\n",
    "    if genome_dir is None:\n",
    "        tempdir = tempfile.TemporaryDirectory()\n",
    "        genome_dir = tempdir.name\n",
    "\n",
    "    if install_genome:\n",
    "        g = genomepy.install_genome(genome_name, genome_provider, genomes_dir=genome_dir)\n",
    "    else:\n",
    "        g = genomepy.Genome(genome_name, genomes_dir=genome_dir)\n",
    "\n",
    "    chroms = adata.var[chr_var_key].unique()\n",
    "    df = adata.var[[chr_var_key, start_var_key, end_var_key]]\n",
    "    seq_dfs = []\n",
    "\n",
    "    for chrom in chroms:\n",
    "        chrom_df = df[df[chr_var_key] == chrom]\n",
    "        block_mid = (chrom_df[start_var_key] + chrom_df[end_var_key]) // 2\n",
    "        block_starts = block_mid - (seq_len // 2)\n",
    "        block_ends = block_starts + seq_len\n",
    "        seqs = []\n",
    "\n",
    "        for start, end in zip(block_starts, block_ends - 1):\n",
    "            seq = str(g.get_seq(chrom, start, end)).upper()\n",
    "            seqs.append(list(seq))\n",
    "\n",
    "        assert len(seqs) == len(chrom_df)\n",
    "        seq_dfs.append(pd.DataFrame(seqs, index=chrom_df.index))\n",
    "\n",
    "    sequence_df = pd.concat(seq_dfs, axis=0).loc[adata.var_names]\n",
    "    adata.varm[sequence_varm_key] = sequence_df\n",
    "    adata.varm[code_varm_key] = sequence_df.applymap(_dna_to_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1df5ca311ae749edba2390e36ff43685",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0.00/159M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/anndata/_core/anndata.py:1830: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added `interval` annotation for features from data/multiome10k.h5mu\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "url = \"https://cf.10xgenomics.com/samples/cell-arc/2.0.0/10k_PBMC_Multiome_nextgem_Chromium_X/10k_PBMC_Multiome_nextgem_Chromium_X_filtered_feature_bc_matrix.h5\"\n",
    "mdata = muon.read_10x_h5(\"data/multiome10k.h5mu\", backup_url=url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = mdata.mod[\"atac\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10970, 111743)\n",
      "(10970, 37054)\n"
     ]
    }
   ],
   "source": [
    "print(adata.shape)\n",
    "# compute the threshold: 5% of the cells\n",
    "min_cells = int(adata.shape[0] * 0.05)\n",
    "# in-place filtering of regions\n",
    "sc.pp.filter_genes(adata, min_cells=min_cells)\n",
    "print(adata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>1422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>4536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>5981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>3150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>6830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>6247</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37054 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        gene_ids feature_types  genome  \\\n",
       "chr1:629395-630394            chr1:629395-630394         Peaks  GRCh38   \n",
       "chr1:633578-634591            chr1:633578-634591         Peaks  GRCh38   \n",
       "chr1:778283-779200            chr1:778283-779200         Peaks  GRCh38   \n",
       "chr1:816873-817775            chr1:816873-817775         Peaks  GRCh38   \n",
       "chr1:827067-827949            chr1:827067-827949         Peaks  GRCh38   \n",
       "...                                          ...           ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583         Peaks  GRCh38   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446         Peaks  GRCh38   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169         Peaks  GRCh38   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332         Peaks  GRCh38   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374         Peaks  GRCh38   \n",
       "\n",
       "                                        interval  n_cells  \n",
       "chr1:629395-630394            chr1:629395-630394     1422  \n",
       "chr1:633578-634591            chr1:633578-634591     4536  \n",
       "chr1:778283-779200            chr1:778283-779200     5981  \n",
       "chr1:816873-817775            chr1:816873-817775      564  \n",
       "chr1:827067-827949            chr1:827067-827949     3150  \n",
       "...                                          ...      ...  \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583      781  \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446      639  \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169     6830  \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332      605  \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374     6247  \n",
       "\n",
       "[37054 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>interval</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:629395-630394</td>\n",
       "      <td>1422</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629395</td>\n",
       "      <td>630394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:633578-634591</td>\n",
       "      <td>4536</td>\n",
       "      <td>chr1</td>\n",
       "      <td>633578</td>\n",
       "      <td>634591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:778283-779200</td>\n",
       "      <td>5981</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778283</td>\n",
       "      <td>779200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:816873-817775</td>\n",
       "      <td>564</td>\n",
       "      <td>chr1</td>\n",
       "      <td>816873</td>\n",
       "      <td>817775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chr1:827067-827949</td>\n",
       "      <td>3150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>827067</td>\n",
       "      <td>827949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:44739-45583</th>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:44739-45583</td>\n",
       "      <td>781</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>44739</td>\n",
       "      <td>45583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:45726-46446</th>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:45726-46446</td>\n",
       "      <td>639</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>45726</td>\n",
       "      <td>46446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GL000219.1:99267-100169</th>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>GL000219.1:99267-100169</td>\n",
       "      <td>6830</td>\n",
       "      <td>GL000219.1</td>\n",
       "      <td>99267</td>\n",
       "      <td>100169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270726.1:41483-42332</th>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270726.1:41483-42332</td>\n",
       "      <td>605</td>\n",
       "      <td>KI270726.1</td>\n",
       "      <td>41483</td>\n",
       "      <td>42332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KI270713.1:21453-22374</th>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>Peaks</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KI270713.1:21453-22374</td>\n",
       "      <td>6247</td>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>21453</td>\n",
       "      <td>22374</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37054 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        gene_ids feature_types  genome  \\\n",
       "chr1:629395-630394            chr1:629395-630394         Peaks  GRCh38   \n",
       "chr1:633578-634591            chr1:633578-634591         Peaks  GRCh38   \n",
       "chr1:778283-779200            chr1:778283-779200         Peaks  GRCh38   \n",
       "chr1:816873-817775            chr1:816873-817775         Peaks  GRCh38   \n",
       "chr1:827067-827949            chr1:827067-827949         Peaks  GRCh38   \n",
       "...                                          ...           ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583         Peaks  GRCh38   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446         Peaks  GRCh38   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169         Peaks  GRCh38   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332         Peaks  GRCh38   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374         Peaks  GRCh38   \n",
       "\n",
       "                                        interval  n_cells         chr   start  \\\n",
       "chr1:629395-630394            chr1:629395-630394     1422        chr1  629395   \n",
       "chr1:633578-634591            chr1:633578-634591     4536        chr1  633578   \n",
       "chr1:778283-779200            chr1:778283-779200     5981        chr1  778283   \n",
       "chr1:816873-817775            chr1:816873-817775      564        chr1  816873   \n",
       "chr1:827067-827949            chr1:827067-827949     3150        chr1  827067   \n",
       "...                                          ...      ...         ...     ...   \n",
       "GL000219.1:44739-45583    GL000219.1:44739-45583      781  GL000219.1   44739   \n",
       "GL000219.1:45726-46446    GL000219.1:45726-46446      639  GL000219.1   45726   \n",
       "GL000219.1:99267-100169  GL000219.1:99267-100169     6830  GL000219.1   99267   \n",
       "KI270726.1:41483-42332    KI270726.1:41483-42332      605  KI270726.1   41483   \n",
       "KI270713.1:21453-22374    KI270713.1:21453-22374     6247  KI270713.1   21453   \n",
       "\n",
       "                            end  \n",
       "chr1:629395-630394       630394  \n",
       "chr1:633578-634591       634591  \n",
       "chr1:778283-779200       779200  \n",
       "chr1:816873-817775       817775  \n",
       "chr1:827067-827949       827949  \n",
       "...                         ...  \n",
       "GL000219.1:44739-45583    45583  \n",
       "GL000219.1:45726-46446    46446  \n",
       "GL000219.1:99267-100169  100169  \n",
       "KI270726.1:41483-42332    42332  \n",
       "KI270713.1:21453-22374    22374  \n",
       "\n",
       "[37054 rows x 8 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "split_interval = adata.var[\"gene_ids\"].str.split(\":\", expand=True)\n",
    "adata.var[\"chr\"] = split_interval[0]\n",
    "split_start_end = split_interval[1].str.split(\"-\", expand=True)\n",
    "adata.var[\"start\"] = split_start_end[0].astype(int)\n",
    "adata.var[\"end\"] = split_start_end[1].astype(int)\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out non-chromosomal regions\n",
    "mask = adata.var[\"chr\"].str.startswith(\"chr\")\n",
    "adata = adata[:, mask].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m08:47:48\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m Downloading assembly summaries from GENCODE\n",
      "\u001b[32m08:48:18\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m Downloading assembly summaries from UCSC\n",
      "\u001b[32m08:48:19\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m Downloading genome from GENCODE. Target URL: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "234f911c88154ea9992f67681bb19012",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Download:   0%|          | 0.00/938M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m08:48:29\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m Genome download successful, starting post processing...\n",
      "\u001b[32m08:48:43\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m name: hg38\n",
      "\u001b[32m08:48:43\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m local name: GRCh38\n",
      "\u001b[32m08:48:43\u001b[0m \u001b[1m|\u001b[0m \u001b[34mINFO\u001b[0m \u001b[1m|\u001b[0m fasta: /cellar/users/aklie/data/GRCh38/GRCh38.fa\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c1d53778e371499194ad2a58430b9c95",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Filtering Fasta: 0.00 lines [00:00, ? lines/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 10970 × 37042\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'interval', 'n_cells', 'chr', 'start', 'end'\n",
       "    varm: 'dna_sequence', 'dna_code'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_dna_sequence(\n",
    "    adata,\n",
    "    genome_name=\"GRCh38\",\n",
    "    genome_dir=\"data\",\n",
    "    chr_var_key=\"chr\",\n",
    "    start_var_key=\"start\",\n",
    "    end_var_key=\"end\",\n",
    ")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1334</th>\n",
       "      <th>1335</th>\n",
       "      <th>1336</th>\n",
       "      <th>1337</th>\n",
       "      <th>1338</th>\n",
       "      <th>1339</th>\n",
       "      <th>1340</th>\n",
       "      <th>1341</th>\n",
       "      <th>1342</th>\n",
       "      <th>1343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:19077075-19078016</th>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:19567013-19567787</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:19744368-19745303</th>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:20575244-20576162</th>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:21240021-21240909</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37042 rows × 1344 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0    1    2    3    4    5    6    7    8    9     ...  \\\n",
       "chr1:629395-630394        C    A    C    T    C    T    C    C    C    C  ...   \n",
       "chr1:633578-634591        G    A    A    A    T    A    G    G    G    C  ...   \n",
       "chr1:778283-779200        C    G    C    C    C    G    G    C    T    A  ...   \n",
       "chr1:816873-817775        A    A    T    T    C    A    T    A    T    G  ...   \n",
       "chr1:827067-827949        C    T    C    T    C    C    T    G    C    C  ...   \n",
       "...                     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "chrY:19077075-19078016    A    C    G    A    C    C    T    C    C    C  ...   \n",
       "chrY:19567013-19567787    G    G    A    G    T    C    T    G    G    G  ...   \n",
       "chrY:19744368-19745303    T    A    T    T    T    T    T    G    T    C  ...   \n",
       "chrY:20575244-20576162    T    T    T    A    C    T    G    T    C    T  ...   \n",
       "chrY:21240021-21240909    A    T    T    G    G    C    C    C    C    C  ...   \n",
       "\n",
       "                       1334 1335 1336 1337 1338 1339 1340 1341 1342 1343  \n",
       "chr1:629395-630394        C    T    A    T    A    T    C    T    A    A  \n",
       "chr1:633578-634591        T    A    A    A    T    C    C    C    C    T  \n",
       "chr1:778283-779200        G    A    C    A    G    G    A    G    T    T  \n",
       "chr1:816873-817775        T    T    A    G    C    G    G    C    T    G  \n",
       "chr1:827067-827949        C    G    T    T    A    T    T    A    A    T  \n",
       "...                     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  \n",
       "chrY:19077075-19078016    C    A    T    A    G    T    T    C    T    A  \n",
       "chrY:19567013-19567787    T    C    T    C    T    T    C    G    T    T  \n",
       "chrY:19744368-19745303    A    T    G    T    G    G    A    A    A    T  \n",
       "chrY:20575244-20576162    G    A    G    T    G    T    A    A    C    A  \n",
       "chrY:21240021-21240909    C    C    T    T    T    C    T    G    A    G  \n",
       "\n",
       "[37042 rows x 1344 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.varm[\"dna_sequence\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1334</th>\n",
       "      <th>1335</th>\n",
       "      <th>1336</th>\n",
       "      <th>1337</th>\n",
       "      <th>1338</th>\n",
       "      <th>1339</th>\n",
       "      <th>1340</th>\n",
       "      <th>1341</th>\n",
       "      <th>1342</th>\n",
       "      <th>1343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:629395-630394</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:633578-634591</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:778283-779200</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:816873-817775</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:827067-827949</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:19077075-19078016</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:19567013-19567787</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:19744368-19745303</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:20575244-20576162</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY:21240021-21240909</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37042 rows × 1344 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        0     1     2     3     4     5     6     7     8     \\\n",
       "chr1:629395-630394         1     0     1     3     1     3     1     1     1   \n",
       "chr1:633578-634591         2     0     0     0     3     0     2     2     2   \n",
       "chr1:778283-779200         1     2     1     1     1     2     2     1     3   \n",
       "chr1:816873-817775         0     0     3     3     1     0     3     0     3   \n",
       "chr1:827067-827949         1     3     1     3     1     1     3     2     1   \n",
       "...                      ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "chrY:19077075-19078016     0     1     2     0     1     1     3     1     1   \n",
       "chrY:19567013-19567787     2     2     0     2     3     1     3     2     2   \n",
       "chrY:19744368-19745303     3     0     3     3     3     3     3     2     3   \n",
       "chrY:20575244-20576162     3     3     3     0     1     3     2     3     1   \n",
       "chrY:21240021-21240909     0     3     3     2     2     1     1     1     1   \n",
       "\n",
       "                        9     ...  1334  1335  1336  1337  1338  1339  1340  \\\n",
       "chr1:629395-630394         1  ...     1     3     0     3     0     3     1   \n",
       "chr1:633578-634591         1  ...     3     0     0     0     3     1     1   \n",
       "chr1:778283-779200         0  ...     2     0     1     0     2     2     0   \n",
       "chr1:816873-817775         2  ...     3     3     0     2     1     2     2   \n",
       "chr1:827067-827949         1  ...     1     2     3     3     0     3     3   \n",
       "...                      ...  ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "chrY:19077075-19078016     1  ...     1     0     3     0     2     3     3   \n",
       "chrY:19567013-19567787     2  ...     3     1     3     1     3     3     1   \n",
       "chrY:19744368-19745303     1  ...     0     3     2     3     2     2     0   \n",
       "chrY:20575244-20576162     3  ...     2     0     2     3     2     3     0   \n",
       "chrY:21240021-21240909     1  ...     1     1     3     3     3     1     3   \n",
       "\n",
       "                        1341  1342  1343  \n",
       "chr1:629395-630394         3     0     0  \n",
       "chr1:633578-634591         1     1     3  \n",
       "chr1:778283-779200         2     3     3  \n",
       "chr1:816873-817775         1     3     2  \n",
       "chr1:827067-827949         0     0     3  \n",
       "...                      ...   ...   ...  \n",
       "chrY:19077075-19078016     1     3     0  \n",
       "chrY:19567013-19567787     2     3     3  \n",
       "chrY:19744368-19745303     0     0     3  \n",
       "chrY:20575244-20576162     0     1     0  \n",
       "chrY:21240021-21240909     2     0     2  \n",
       "\n",
       "[37042 rows x 1344 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.varm[\"dna_code\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "bdata = adata.transpose()\n",
    "bdata.layers[\"binary\"] = (bdata.X.copy() > 0).astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be22bd02a2ff4c06ba66670a859278bd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0.00/5.58M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2700 × 32738\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.datasets.pbmc3k()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['size_factors'] = adata.X.sum(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "use_cuda = torch.cuda.is_available()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataloader = AnnLoader(adata, batch_size=128, shuffle=True, use_cuda=use_cuda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnCollection object with n_obs × n_vars = 2700 × 32738\n",
       "  constructed from 1 AnnData objects\n",
       "    obs: 'size_factors'"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataloader.dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnCollectionView object with n_obs × n_vars = 10 × 32738\n",
       "    obs: 'size_factors'"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataloader.dataset[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X: cuda:0 torch.float32\n"
     ]
    }
   ],
   "source": [
    "batch = dataloader.dataset[:10]\n",
    "\n",
    "print('X:', batch.X.device, batch.X.dtype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([10, 32738])"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/zarr/creation.py:300: UserWarning: ignoring keyword argument 'maxshape'\n",
      "  warn(\"ignoring keyword argument %r\" % k)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<zarr.hierarchy.Group '/'>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.write_zarr(\"/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\")\n",
    "zarr.consolidate_metadata(\"/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<zarr.hierarchy.Group '/'>"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zarr.open_consolidated(\"/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2638 × 1838\n",
       "    obs: 'n_genes', 'percent_mito', 'n_counts', 'louvain'\n",
       "    var: 'n_cells'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.read_zarr(\"/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;31mSignature:\u001b[0m\n",
      "\u001b[0mxr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen_zarr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mstore\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mgroup\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0msynchronizer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mchunks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'auto'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdecode_cf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mmask_and_scale\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdecode_times\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mconcat_characters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdecode_coords\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdrop_variables\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mconsolidated\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0moverwrite_encoded_chunks\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mchunk_store\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mstorage_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mdecode_timedelta\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0muse_cftime\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mzarr_version\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mchunked_array_type\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'str | None'\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0mfrom_array_kwargs\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'dict[str, Any] | None'\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m    \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\n",
      "\u001b[0;34m\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mDocstring:\u001b[0m\n",
      "Load and decode a dataset from a Zarr store.\n",
      "\n",
      "The `store` object should be a valid store for a Zarr group. `store`\n",
      "variables must contain dimension metadata encoded in the\n",
      "`_ARRAY_DIMENSIONS` attribute or must have NCZarr format.\n",
      "\n",
      "Parameters\n",
      "----------\n",
      "store : MutableMapping or str\n",
      "    A MutableMapping where a Zarr Group has been stored or a path to a\n",
      "    directory in file system where a Zarr DirectoryStore has been stored.\n",
      "synchronizer : object, optional\n",
      "    Array synchronizer provided to zarr\n",
      "group : str, optional\n",
      "    Group path. (a.k.a. `path` in zarr terminology.)\n",
      "chunks : int or dict or tuple or {None, 'auto'}, optional\n",
      "    Chunk sizes along each dimension, e.g., ``5`` or\n",
      "    ``{'x': 5, 'y': 5}``. If `chunks='auto'`, dask chunks are created\n",
      "    based on the variable's zarr chunks. If `chunks=None`, zarr array\n",
      "    data will lazily convert to numpy arrays upon access. This accepts\n",
      "    all the chunk specifications as Dask does.\n",
      "overwrite_encoded_chunks : bool, optional\n",
      "    Whether to drop the zarr chunks encoded for each variable when a\n",
      "    dataset is loaded with specified chunk sizes (default: False)\n",
      "decode_cf : bool, optional\n",
      "    Whether to decode these variables, assuming they were saved according\n",
      "    to CF conventions.\n",
      "mask_and_scale : bool, optional\n",
      "    If True, replace array values equal to `_FillValue` with NA and scale\n",
      "    values according to the formula `original_values * scale_factor +\n",
      "    add_offset`, where `_FillValue`, `scale_factor` and `add_offset` are\n",
      "    taken from variable attributes (if they exist).  If the `_FillValue` or\n",
      "    `missing_value` attribute contains multiple values a warning will be\n",
      "    issued and all array values matching one of the multiple values will\n",
      "    be replaced by NA.\n",
      "decode_times : bool, optional\n",
      "    If True, decode times encoded in the standard NetCDF datetime format\n",
      "    into datetime objects. Otherwise, leave them encoded as numbers.\n",
      "concat_characters : bool, optional\n",
      "    If True, concatenate along the last dimension of character arrays to\n",
      "    form string arrays. Dimensions will only be concatenated over (and\n",
      "    removed) if they have no corresponding variable and if they are only\n",
      "    used as the last dimension of character arrays.\n",
      "decode_coords : bool, optional\n",
      "    If True, decode the 'coordinates' attribute to identify coordinates in\n",
      "    the resulting dataset.\n",
      "drop_variables : str or iterable, optional\n",
      "    A variable or list of variables to exclude from being parsed from the\n",
      "    dataset. This may be useful to drop variables with problems or\n",
      "    inconsistent values.\n",
      "consolidated : bool, optional\n",
      "    Whether to open the store using zarr's consolidated metadata\n",
      "    capability. Only works for stores that have already been consolidated.\n",
      "    By default (`consolidate=None`), attempts to read consolidated metadata,\n",
      "    falling back to read non-consolidated metadata if that fails.\n",
      "\n",
      "    When the experimental ``zarr_version=3``, ``consolidated`` must be\n",
      "    either be ``None`` or ``False``.\n",
      "chunk_store : MutableMapping, optional\n",
      "    A separate Zarr store only for chunk data.\n",
      "storage_options : dict, optional\n",
      "    Any additional parameters for the storage backend (ignored for local\n",
      "    paths).\n",
      "decode_timedelta : bool, optional\n",
      "    If True, decode variables and coordinates with time units in\n",
      "    {'days', 'hours', 'minutes', 'seconds', 'milliseconds', 'microseconds'}\n",
      "    into timedelta objects. If False, leave them encoded as numbers.\n",
      "    If None (default), assume the same value of decode_time.\n",
      "use_cftime : bool, optional\n",
      "    Only relevant if encoded dates come from a standard calendar\n",
      "    (e.g. \"gregorian\", \"proleptic_gregorian\", \"standard\", or not\n",
      "    specified).  If None (default), attempt to decode times to\n",
      "    ``np.datetime64[ns]`` objects; if this is not possible, decode times to\n",
      "    ``cftime.datetime`` objects. If True, always decode times to\n",
      "    ``cftime.datetime`` objects, regardless of whether or not they can be\n",
      "    represented using ``np.datetime64[ns]`` objects.  If False, always\n",
      "    decode times to ``np.datetime64[ns]`` objects; if this is not possible\n",
      "    raise an error.\n",
      "zarr_version : int or None, optional\n",
      "    The desired zarr spec version to target (currently 2 or 3). The default\n",
      "    of None will attempt to determine the zarr version from ``store`` when\n",
      "    possible, otherwise defaulting to 2.\n",
      "chunked_array_type: str, optional\n",
      "    Which chunked array type to coerce this datasets' arrays to.\n",
      "    Defaults to 'dask' if installed, else whatever is registered via the `ChunkManagerEntryPoint` system.\n",
      "    Experimental API that should not be relied upon.\n",
      "from_array_kwargs: dict, optional\n",
      "    Additional keyword arguments passed on to the `ChunkManagerEntrypoint.from_array` method used to create\n",
      "    chunked arrays, via whichever chunk manager is specified through the `chunked_array_type` kwarg.\n",
      "    Defaults to {'manager': 'dask'}, meaning additional kwargs will be passed eventually to\n",
      "    :py:func:`dask.array.from_array`. Experimental API that should not be relied upon.\n",
      "\n",
      "Returns\n",
      "-------\n",
      "dataset : Dataset\n",
      "    The newly created dataset.\n",
      "\n",
      "See Also\n",
      "--------\n",
      "open_dataset\n",
      "open_mfdataset\n",
      "\n",
      "References\n",
      "----------\n",
      "http://zarr.readthedocs.io/\n",
      "\u001b[0;31mFile:\u001b[0m      ~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py\n",
      "\u001b[0;31mType:\u001b[0m      function"
     ]
    }
   ],
   "source": [
    "xr.open_zarr?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "the JSON object must be str, bytes or bytearray, not dict",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:210\u001b[0m, in \u001b[0;36m_get_zarr_dims_and_attrs\u001b[0;34m(zarr_obj, dimension_key, try_nczarr)\u001b[0m\n\u001b[1;32m    208\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    209\u001b[0m     \u001b[39m# Xarray-Zarr\u001b[39;00m\n\u001b[0;32m--> 210\u001b[0m     dimensions \u001b[39m=\u001b[39m zarr_obj\u001b[39m.\u001b[39;49mattrs[dimension_key]\n\u001b[1;32m    211\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/zarr/attrs.py:72\u001b[0m, in \u001b[0;36mAttributes.__getitem__\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__getitem__\u001b[39m(\u001b[39mself\u001b[39m, item):\n\u001b[0;32m---> 72\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49masdict()[item]\n",
      "\u001b[0;31mKeyError\u001b[0m: '_ARRAY_DIMENSIONS'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[59], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m xr\u001b[39m.\u001b[39;49mopen_zarr(\u001b[39m\"\u001b[39;49m\u001b[39m/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:944\u001b[0m, in \u001b[0;36mopen_zarr\u001b[0;34m(store, group, synchronizer, chunks, decode_cf, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, consolidated, overwrite_encoded_chunks, chunk_store, storage_options, decode_timedelta, use_cftime, zarr_version, chunked_array_type, from_array_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m    930\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\n\u001b[1;32m    931\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mopen_zarr() got unexpected keyword arguments \u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m,\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin(kwargs\u001b[39m.\u001b[39mkeys())\n\u001b[1;32m    932\u001b[0m     )\n\u001b[1;32m    934\u001b[0m backend_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    935\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39msynchronizer\u001b[39m\u001b[39m\"\u001b[39m: synchronizer,\n\u001b[1;32m    936\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mconsolidated\u001b[39m\u001b[39m\"\u001b[39m: consolidated,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    941\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mzarr_version\u001b[39m\u001b[39m\"\u001b[39m: zarr_version,\n\u001b[1;32m    942\u001b[0m }\n\u001b[0;32m--> 944\u001b[0m ds \u001b[39m=\u001b[39m open_dataset(\n\u001b[1;32m    945\u001b[0m     filename_or_obj\u001b[39m=\u001b[39;49mstore,\n\u001b[1;32m    946\u001b[0m     group\u001b[39m=\u001b[39;49mgroup,\n\u001b[1;32m    947\u001b[0m     decode_cf\u001b[39m=\u001b[39;49mdecode_cf,\n\u001b[1;32m    948\u001b[0m     mask_and_scale\u001b[39m=\u001b[39;49mmask_and_scale,\n\u001b[1;32m    949\u001b[0m     decode_times\u001b[39m=\u001b[39;49mdecode_times,\n\u001b[1;32m    950\u001b[0m     concat_characters\u001b[39m=\u001b[39;49mconcat_characters,\n\u001b[1;32m    951\u001b[0m     decode_coords\u001b[39m=\u001b[39;49mdecode_coords,\n\u001b[1;32m    952\u001b[0m     engine\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mzarr\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    953\u001b[0m     chunks\u001b[39m=\u001b[39;49mchunks,\n\u001b[1;32m    954\u001b[0m     drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m    955\u001b[0m     chunked_array_type\u001b[39m=\u001b[39;49mchunked_array_type,\n\u001b[1;32m    956\u001b[0m     from_array_kwargs\u001b[39m=\u001b[39;49mfrom_array_kwargs,\n\u001b[1;32m    957\u001b[0m     backend_kwargs\u001b[39m=\u001b[39;49mbackend_kwargs,\n\u001b[1;32m    958\u001b[0m     decode_timedelta\u001b[39m=\u001b[39;49mdecode_timedelta,\n\u001b[1;32m    959\u001b[0m     use_cftime\u001b[39m=\u001b[39;49muse_cftime,\n\u001b[1;32m    960\u001b[0m     zarr_version\u001b[39m=\u001b[39;49mzarr_version,\n\u001b[1;32m    961\u001b[0m )\n\u001b[1;32m    962\u001b[0m \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/api.py:572\u001b[0m, in \u001b[0;36mopen_dataset\u001b[0;34m(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, inline_array, chunked_array_type, from_array_kwargs, backend_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m    560\u001b[0m decoders \u001b[39m=\u001b[39m _resolve_decoders_kwargs(\n\u001b[1;32m    561\u001b[0m     decode_cf,\n\u001b[1;32m    562\u001b[0m     open_backend_dataset_parameters\u001b[39m=\u001b[39mbackend\u001b[39m.\u001b[39mopen_dataset_parameters,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    568\u001b[0m     decode_coords\u001b[39m=\u001b[39mdecode_coords,\n\u001b[1;32m    569\u001b[0m )\n\u001b[1;32m    571\u001b[0m overwrite_encoded_chunks \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(\u001b[39m\"\u001b[39m\u001b[39moverwrite_encoded_chunks\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m)\n\u001b[0;32m--> 572\u001b[0m backend_ds \u001b[39m=\u001b[39m backend\u001b[39m.\u001b[39;49mopen_dataset(\n\u001b[1;32m    573\u001b[0m     filename_or_obj,\n\u001b[1;32m    574\u001b[0m     drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m    575\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mdecoders,\n\u001b[1;32m    576\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m    577\u001b[0m )\n\u001b[1;32m    578\u001b[0m ds \u001b[39m=\u001b[39m _dataset_from_backend_dataset(\n\u001b[1;32m    579\u001b[0m     backend_ds,\n\u001b[1;32m    580\u001b[0m     filename_or_obj,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    590\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs,\n\u001b[1;32m    591\u001b[0m )\n\u001b[1;32m    592\u001b[0m \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:1026\u001b[0m, in \u001b[0;36mZarrBackendEntrypoint.open_dataset\u001b[0;34m(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta, group, mode, synchronizer, consolidated, chunk_store, storage_options, stacklevel, zarr_version)\u001b[0m\n\u001b[1;32m   1024\u001b[0m store_entrypoint \u001b[39m=\u001b[39m StoreBackendEntrypoint()\n\u001b[1;32m   1025\u001b[0m \u001b[39mwith\u001b[39;00m close_on_error(store):\n\u001b[0;32m-> 1026\u001b[0m     ds \u001b[39m=\u001b[39m store_entrypoint\u001b[39m.\u001b[39;49mopen_dataset(\n\u001b[1;32m   1027\u001b[0m         store,\n\u001b[1;32m   1028\u001b[0m         mask_and_scale\u001b[39m=\u001b[39;49mmask_and_scale,\n\u001b[1;32m   1029\u001b[0m         decode_times\u001b[39m=\u001b[39;49mdecode_times,\n\u001b[1;32m   1030\u001b[0m         concat_characters\u001b[39m=\u001b[39;49mconcat_characters,\n\u001b[1;32m   1031\u001b[0m         decode_coords\u001b[39m=\u001b[39;49mdecode_coords,\n\u001b[1;32m   1032\u001b[0m         drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m   1033\u001b[0m         use_cftime\u001b[39m=\u001b[39;49muse_cftime,\n\u001b[1;32m   1034\u001b[0m         decode_timedelta\u001b[39m=\u001b[39;49mdecode_timedelta,\n\u001b[1;32m   1035\u001b[0m     )\n\u001b[1;32m   1036\u001b[0m \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/store.py:43\u001b[0m, in \u001b[0;36mStoreBackendEntrypoint.open_dataset\u001b[0;34m(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta)\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mopen_dataset\u001b[39m(  \u001b[39m# type: ignore[override]  # allow LSP violation, not supporting **kwargs\u001b[39;00m\n\u001b[1;32m     30\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m     31\u001b[0m     filename_or_obj: \u001b[39mstr\u001b[39m \u001b[39m|\u001b[39m os\u001b[39m.\u001b[39mPathLike[Any] \u001b[39m|\u001b[39m BufferedIOBase \u001b[39m|\u001b[39m AbstractDataStore,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     39\u001b[0m     decode_timedelta\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m     40\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Dataset:\n\u001b[1;32m     41\u001b[0m     \u001b[39massert\u001b[39;00m \u001b[39misinstance\u001b[39m(filename_or_obj, AbstractDataStore)\n\u001b[0;32m---> 43\u001b[0m     \u001b[39mvars\u001b[39m, attrs \u001b[39m=\u001b[39m filename_or_obj\u001b[39m.\u001b[39;49mload()\n\u001b[1;32m     44\u001b[0m     encoding \u001b[39m=\u001b[39m filename_or_obj\u001b[39m.\u001b[39mget_encoding()\n\u001b[1;32m     46\u001b[0m     \u001b[39mvars\u001b[39m, attrs, coord_names \u001b[39m=\u001b[39m conventions\u001b[39m.\u001b[39mdecode_cf_variables(\n\u001b[1;32m     47\u001b[0m         \u001b[39mvars\u001b[39m,\n\u001b[1;32m     48\u001b[0m         attrs,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     55\u001b[0m         decode_timedelta\u001b[39m=\u001b[39mdecode_timedelta,\n\u001b[1;32m     56\u001b[0m     )\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/common.py:210\u001b[0m, in \u001b[0;36mAbstractDataStore.load\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mload\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m    189\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    190\u001b[0m \u001b[39m    This loads the variables and attributes simultaneously.\u001b[39;00m\n\u001b[1;32m    191\u001b[0m \u001b[39m    A centralized loading function makes it easier to create\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    207\u001b[0m \u001b[39m    are requested, so care should be taken to make sure its fast.\u001b[39;00m\n\u001b[1;32m    208\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m    209\u001b[0m     variables \u001b[39m=\u001b[39m FrozenDict(\n\u001b[0;32m--> 210\u001b[0m         (_decode_variable_name(k), v) \u001b[39mfor\u001b[39;00m k, v \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_variables()\u001b[39m.\u001b[39mitems()\n\u001b[1;32m    211\u001b[0m     )\n\u001b[1;32m    212\u001b[0m     attributes \u001b[39m=\u001b[39m FrozenDict(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_attrs())\n\u001b[1;32m    213\u001b[0m     \u001b[39mreturn\u001b[39;00m variables, attributes\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:533\u001b[0m, in \u001b[0;36mZarrStore.get_variables\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    532\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_variables\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 533\u001b[0m     \u001b[39mreturn\u001b[39;00m FrozenDict(\n\u001b[1;32m    534\u001b[0m         (k, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mopen_store_variable(k, v)) \u001b[39mfor\u001b[39;49;00m k, v \u001b[39min\u001b[39;49;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mzarr_group\u001b[39m.\u001b[39;49marrays()\n\u001b[1;32m    535\u001b[0m     )\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/core/utils.py:476\u001b[0m, in \u001b[0;36mFrozenDict\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    475\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mFrozenDict\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Frozen:\n\u001b[0;32m--> 476\u001b[0m     \u001b[39mreturn\u001b[39;00m Frozen(\u001b[39mdict\u001b[39;49m(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs))\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:534\u001b[0m, in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    532\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_variables\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m    533\u001b[0m     \u001b[39mreturn\u001b[39;00m FrozenDict(\n\u001b[0;32m--> 534\u001b[0m         (k, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mopen_store_variable(k, v)) \u001b[39mfor\u001b[39;00m k, v \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mzarr_group\u001b[39m.\u001b[39marrays()\n\u001b[1;32m    535\u001b[0m     )\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:510\u001b[0m, in \u001b[0;36mZarrStore.open_store_variable\u001b[0;34m(self, name, zarr_array)\u001b[0m\n\u001b[1;32m    508\u001b[0m data \u001b[39m=\u001b[39m indexing\u001b[39m.\u001b[39mLazilyIndexedArray(ZarrArrayWrapper(zarr_array))\n\u001b[1;32m    509\u001b[0m try_nczarr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_mode \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 510\u001b[0m dimensions, attributes \u001b[39m=\u001b[39m _get_zarr_dims_and_attrs(\n\u001b[1;32m    511\u001b[0m     zarr_array, DIMENSION_KEY, try_nczarr\n\u001b[1;32m    512\u001b[0m )\n\u001b[1;32m    513\u001b[0m attributes \u001b[39m=\u001b[39m \u001b[39mdict\u001b[39m(attributes)\n\u001b[1;32m    515\u001b[0m \u001b[39m# TODO: this should not be needed once\u001b[39;00m\n\u001b[1;32m    516\u001b[0m \u001b[39m# https://github.com/zarr-developers/zarr-python/issues/1269 is resolved.\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:220\u001b[0m, in \u001b[0;36m_get_zarr_dims_and_attrs\u001b[0;34m(zarr_obj, dimension_key, try_nczarr)\u001b[0m\n\u001b[1;32m    218\u001b[0m \u001b[39m# NCZarr defines dimensions through metadata in .zarray\u001b[39;00m\n\u001b[1;32m    219\u001b[0m zarray_path \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mjoin(zarr_obj\u001b[39m.\u001b[39mpath, \u001b[39m\"\u001b[39m\u001b[39m.zarray\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 220\u001b[0m zarray \u001b[39m=\u001b[39m json\u001b[39m.\u001b[39;49mloads(zarr_obj\u001b[39m.\u001b[39;49mstore[zarray_path])\n\u001b[1;32m    221\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    222\u001b[0m     \u001b[39m# NCZarr uses Fully Qualified Names\u001b[39;00m\n\u001b[1;32m    223\u001b[0m     dimensions \u001b[39m=\u001b[39m [\n\u001b[1;32m    224\u001b[0m         os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mbasename(dim) \u001b[39mfor\u001b[39;00m dim \u001b[39min\u001b[39;00m zarray[\u001b[39m\"\u001b[39m\u001b[39m_NCZARR_ARRAY\u001b[39m\u001b[39m\"\u001b[39m][\u001b[39m\"\u001b[39m\u001b[39mdimrefs\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[1;32m    225\u001b[0m     ]\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/json/__init__.py:339\u001b[0m, in \u001b[0;36mloads\u001b[0;34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[0m\n\u001b[1;32m    337\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    338\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(s, (\u001b[39mbytes\u001b[39m, \u001b[39mbytearray\u001b[39m)):\n\u001b[0;32m--> 339\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mthe JSON object must be str, bytes or bytearray, \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    340\u001b[0m                         \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mnot \u001b[39m\u001b[39m{\u001b[39;00ms\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m    341\u001b[0m     s \u001b[39m=\u001b[39m s\u001b[39m.\u001b[39mdecode(detect_encoding(s), \u001b[39m'\u001b[39m\u001b[39msurrogatepass\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m    343\u001b[0m \u001b[39mif\u001b[39;00m (\u001b[39mcls\u001b[39m \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m object_hook \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    344\u001b[0m         parse_int \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m parse_float \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    345\u001b[0m         parse_constant \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m object_pairs_hook \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m kw):\n",
      "\u001b[0;31mTypeError\u001b[0m: the JSON object must be str, bytes or bytearray, not dict"
     ]
    }
   ],
   "source": [
    "xr.open_zarr(\"/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "the JSON object must be str, bytes or bytearray, not dict",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:210\u001b[0m, in \u001b[0;36m_get_zarr_dims_and_attrs\u001b[0;34m(zarr_obj, dimension_key, try_nczarr)\u001b[0m\n\u001b[1;32m    208\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    209\u001b[0m     \u001b[39m# Xarray-Zarr\u001b[39;00m\n\u001b[0;32m--> 210\u001b[0m     dimensions \u001b[39m=\u001b[39m zarr_obj\u001b[39m.\u001b[39;49mattrs[dimension_key]\n\u001b[1;32m    211\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/zarr/attrs.py:72\u001b[0m, in \u001b[0;36mAttributes.__getitem__\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__getitem__\u001b[39m(\u001b[39mself\u001b[39m, item):\n\u001b[0;32m---> 72\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49masdict()[item]\n",
      "\u001b[0;31mKeyError\u001b[0m: '_ARRAY_DIMENSIONS'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[51], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m sd\u001b[39m.\u001b[39;49mopen_zarr(\u001b[39m\"\u001b[39;49m\u001b[39m/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m~/projects/ML4GLand/SeqData/seqdata/xarray/seqdata.py:101\u001b[0m, in \u001b[0;36mopen_zarr\u001b[0;34m(store, group, synchronizer, chunks, decode_cf, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, consolidated, overwrite_encoded_chunks, chunk_store, storage_options, decode_timedelta, use_cftime, zarr_version, **kwargs)\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mopen_zarr\u001b[39m(\n\u001b[1;32m     36\u001b[0m     store: PathType,\n\u001b[1;32m     37\u001b[0m     group: Optional[\u001b[39mstr\u001b[39m] \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     55\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs,\n\u001b[1;32m     56\u001b[0m ):\n\u001b[1;32m     57\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Open a SeqData object from disk.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \n\u001b[1;32m     59\u001b[0m \u001b[39m    Parameters\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[39m        SeqData object\u001b[39;00m\n\u001b[1;32m    100\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 101\u001b[0m     ds \u001b[39m=\u001b[39m xr\u001b[39m.\u001b[39;49mopen_zarr(\n\u001b[1;32m    102\u001b[0m         store\u001b[39m=\u001b[39;49mstore,\n\u001b[1;32m    103\u001b[0m         group\u001b[39m=\u001b[39;49mgroup,\n\u001b[1;32m    104\u001b[0m         synchronizer\u001b[39m=\u001b[39;49msynchronizer,\n\u001b[1;32m    105\u001b[0m         chunks\u001b[39m=\u001b[39;49mchunks,  \u001b[39m# type: ignore\u001b[39;49;00m\n\u001b[1;32m    106\u001b[0m         decode_cf\u001b[39m=\u001b[39;49mdecode_cf,\n\u001b[1;32m    107\u001b[0m         mask_and_scale\u001b[39m=\u001b[39;49mmask_and_scale,\n\u001b[1;32m    108\u001b[0m         decode_times\u001b[39m=\u001b[39;49mdecode_times,\n\u001b[1;32m    109\u001b[0m         concat_characters\u001b[39m=\u001b[39;49mconcat_characters,\n\u001b[1;32m    110\u001b[0m         decode_coords\u001b[39m=\u001b[39;49mdecode_coords,\n\u001b[1;32m    111\u001b[0m         drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m    112\u001b[0m         consolidated\u001b[39m=\u001b[39;49mconsolidated,\n\u001b[1;32m    113\u001b[0m         overwrite_encoded_chunks\u001b[39m=\u001b[39;49moverwrite_encoded_chunks,\n\u001b[1;32m    114\u001b[0m         chunk_store\u001b[39m=\u001b[39;49mchunk_store,\n\u001b[1;32m    115\u001b[0m         storage_options\u001b[39m=\u001b[39;49mstorage_options,\n\u001b[1;32m    116\u001b[0m         decode_timedelta\u001b[39m=\u001b[39;49mdecode_timedelta,\n\u001b[1;32m    117\u001b[0m         use_cftime\u001b[39m=\u001b[39;49muse_cftime,\n\u001b[1;32m    118\u001b[0m         zarr_version\u001b[39m=\u001b[39;49mzarr_version,\n\u001b[1;32m    119\u001b[0m         \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m    120\u001b[0m     )\n\u001b[1;32m    121\u001b[0m     \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:944\u001b[0m, in \u001b[0;36mopen_zarr\u001b[0;34m(store, group, synchronizer, chunks, decode_cf, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, consolidated, overwrite_encoded_chunks, chunk_store, storage_options, decode_timedelta, use_cftime, zarr_version, chunked_array_type, from_array_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m    930\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\n\u001b[1;32m    931\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mopen_zarr() got unexpected keyword arguments \u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m,\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mjoin(kwargs\u001b[39m.\u001b[39mkeys())\n\u001b[1;32m    932\u001b[0m     )\n\u001b[1;32m    934\u001b[0m backend_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    935\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39msynchronizer\u001b[39m\u001b[39m\"\u001b[39m: synchronizer,\n\u001b[1;32m    936\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mconsolidated\u001b[39m\u001b[39m\"\u001b[39m: consolidated,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    941\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mzarr_version\u001b[39m\u001b[39m\"\u001b[39m: zarr_version,\n\u001b[1;32m    942\u001b[0m }\n\u001b[0;32m--> 944\u001b[0m ds \u001b[39m=\u001b[39m open_dataset(\n\u001b[1;32m    945\u001b[0m     filename_or_obj\u001b[39m=\u001b[39;49mstore,\n\u001b[1;32m    946\u001b[0m     group\u001b[39m=\u001b[39;49mgroup,\n\u001b[1;32m    947\u001b[0m     decode_cf\u001b[39m=\u001b[39;49mdecode_cf,\n\u001b[1;32m    948\u001b[0m     mask_and_scale\u001b[39m=\u001b[39;49mmask_and_scale,\n\u001b[1;32m    949\u001b[0m     decode_times\u001b[39m=\u001b[39;49mdecode_times,\n\u001b[1;32m    950\u001b[0m     concat_characters\u001b[39m=\u001b[39;49mconcat_characters,\n\u001b[1;32m    951\u001b[0m     decode_coords\u001b[39m=\u001b[39;49mdecode_coords,\n\u001b[1;32m    952\u001b[0m     engine\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mzarr\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m    953\u001b[0m     chunks\u001b[39m=\u001b[39;49mchunks,\n\u001b[1;32m    954\u001b[0m     drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m    955\u001b[0m     chunked_array_type\u001b[39m=\u001b[39;49mchunked_array_type,\n\u001b[1;32m    956\u001b[0m     from_array_kwargs\u001b[39m=\u001b[39;49mfrom_array_kwargs,\n\u001b[1;32m    957\u001b[0m     backend_kwargs\u001b[39m=\u001b[39;49mbackend_kwargs,\n\u001b[1;32m    958\u001b[0m     decode_timedelta\u001b[39m=\u001b[39;49mdecode_timedelta,\n\u001b[1;32m    959\u001b[0m     use_cftime\u001b[39m=\u001b[39;49muse_cftime,\n\u001b[1;32m    960\u001b[0m     zarr_version\u001b[39m=\u001b[39;49mzarr_version,\n\u001b[1;32m    961\u001b[0m )\n\u001b[1;32m    962\u001b[0m \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/api.py:572\u001b[0m, in \u001b[0;36mopen_dataset\u001b[0;34m(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, inline_array, chunked_array_type, from_array_kwargs, backend_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m    560\u001b[0m decoders \u001b[39m=\u001b[39m _resolve_decoders_kwargs(\n\u001b[1;32m    561\u001b[0m     decode_cf,\n\u001b[1;32m    562\u001b[0m     open_backend_dataset_parameters\u001b[39m=\u001b[39mbackend\u001b[39m.\u001b[39mopen_dataset_parameters,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    568\u001b[0m     decode_coords\u001b[39m=\u001b[39mdecode_coords,\n\u001b[1;32m    569\u001b[0m )\n\u001b[1;32m    571\u001b[0m overwrite_encoded_chunks \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(\u001b[39m\"\u001b[39m\u001b[39moverwrite_encoded_chunks\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m)\n\u001b[0;32m--> 572\u001b[0m backend_ds \u001b[39m=\u001b[39m backend\u001b[39m.\u001b[39;49mopen_dataset(\n\u001b[1;32m    573\u001b[0m     filename_or_obj,\n\u001b[1;32m    574\u001b[0m     drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m    575\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mdecoders,\n\u001b[1;32m    576\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m    577\u001b[0m )\n\u001b[1;32m    578\u001b[0m ds \u001b[39m=\u001b[39m _dataset_from_backend_dataset(\n\u001b[1;32m    579\u001b[0m     backend_ds,\n\u001b[1;32m    580\u001b[0m     filename_or_obj,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    590\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs,\n\u001b[1;32m    591\u001b[0m )\n\u001b[1;32m    592\u001b[0m \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:1026\u001b[0m, in \u001b[0;36mZarrBackendEntrypoint.open_dataset\u001b[0;34m(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta, group, mode, synchronizer, consolidated, chunk_store, storage_options, stacklevel, zarr_version)\u001b[0m\n\u001b[1;32m   1024\u001b[0m store_entrypoint \u001b[39m=\u001b[39m StoreBackendEntrypoint()\n\u001b[1;32m   1025\u001b[0m \u001b[39mwith\u001b[39;00m close_on_error(store):\n\u001b[0;32m-> 1026\u001b[0m     ds \u001b[39m=\u001b[39m store_entrypoint\u001b[39m.\u001b[39;49mopen_dataset(\n\u001b[1;32m   1027\u001b[0m         store,\n\u001b[1;32m   1028\u001b[0m         mask_and_scale\u001b[39m=\u001b[39;49mmask_and_scale,\n\u001b[1;32m   1029\u001b[0m         decode_times\u001b[39m=\u001b[39;49mdecode_times,\n\u001b[1;32m   1030\u001b[0m         concat_characters\u001b[39m=\u001b[39;49mconcat_characters,\n\u001b[1;32m   1031\u001b[0m         decode_coords\u001b[39m=\u001b[39;49mdecode_coords,\n\u001b[1;32m   1032\u001b[0m         drop_variables\u001b[39m=\u001b[39;49mdrop_variables,\n\u001b[1;32m   1033\u001b[0m         use_cftime\u001b[39m=\u001b[39;49muse_cftime,\n\u001b[1;32m   1034\u001b[0m         decode_timedelta\u001b[39m=\u001b[39;49mdecode_timedelta,\n\u001b[1;32m   1035\u001b[0m     )\n\u001b[1;32m   1036\u001b[0m \u001b[39mreturn\u001b[39;00m ds\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/store.py:43\u001b[0m, in \u001b[0;36mStoreBackendEntrypoint.open_dataset\u001b[0;34m(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta)\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mopen_dataset\u001b[39m(  \u001b[39m# type: ignore[override]  # allow LSP violation, not supporting **kwargs\u001b[39;00m\n\u001b[1;32m     30\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m     31\u001b[0m     filename_or_obj: \u001b[39mstr\u001b[39m \u001b[39m|\u001b[39m os\u001b[39m.\u001b[39mPathLike[Any] \u001b[39m|\u001b[39m BufferedIOBase \u001b[39m|\u001b[39m AbstractDataStore,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     39\u001b[0m     decode_timedelta\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m     40\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Dataset:\n\u001b[1;32m     41\u001b[0m     \u001b[39massert\u001b[39;00m \u001b[39misinstance\u001b[39m(filename_or_obj, AbstractDataStore)\n\u001b[0;32m---> 43\u001b[0m     \u001b[39mvars\u001b[39m, attrs \u001b[39m=\u001b[39m filename_or_obj\u001b[39m.\u001b[39;49mload()\n\u001b[1;32m     44\u001b[0m     encoding \u001b[39m=\u001b[39m filename_or_obj\u001b[39m.\u001b[39mget_encoding()\n\u001b[1;32m     46\u001b[0m     \u001b[39mvars\u001b[39m, attrs, coord_names \u001b[39m=\u001b[39m conventions\u001b[39m.\u001b[39mdecode_cf_variables(\n\u001b[1;32m     47\u001b[0m         \u001b[39mvars\u001b[39m,\n\u001b[1;32m     48\u001b[0m         attrs,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     55\u001b[0m         decode_timedelta\u001b[39m=\u001b[39mdecode_timedelta,\n\u001b[1;32m     56\u001b[0m     )\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/common.py:210\u001b[0m, in \u001b[0;36mAbstractDataStore.load\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mload\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m    189\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    190\u001b[0m \u001b[39m    This loads the variables and attributes simultaneously.\u001b[39;00m\n\u001b[1;32m    191\u001b[0m \u001b[39m    A centralized loading function makes it easier to create\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    207\u001b[0m \u001b[39m    are requested, so care should be taken to make sure its fast.\u001b[39;00m\n\u001b[1;32m    208\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m    209\u001b[0m     variables \u001b[39m=\u001b[39m FrozenDict(\n\u001b[0;32m--> 210\u001b[0m         (_decode_variable_name(k), v) \u001b[39mfor\u001b[39;00m k, v \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_variables()\u001b[39m.\u001b[39mitems()\n\u001b[1;32m    211\u001b[0m     )\n\u001b[1;32m    212\u001b[0m     attributes \u001b[39m=\u001b[39m FrozenDict(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_attrs())\n\u001b[1;32m    213\u001b[0m     \u001b[39mreturn\u001b[39;00m variables, attributes\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:533\u001b[0m, in \u001b[0;36mZarrStore.get_variables\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    532\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_variables\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 533\u001b[0m     \u001b[39mreturn\u001b[39;00m FrozenDict(\n\u001b[1;32m    534\u001b[0m         (k, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mopen_store_variable(k, v)) \u001b[39mfor\u001b[39;49;00m k, v \u001b[39min\u001b[39;49;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mzarr_group\u001b[39m.\u001b[39;49marrays()\n\u001b[1;32m    535\u001b[0m     )\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/core/utils.py:476\u001b[0m, in \u001b[0;36mFrozenDict\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    475\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mFrozenDict\u001b[39m(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Frozen:\n\u001b[0;32m--> 476\u001b[0m     \u001b[39mreturn\u001b[39;00m Frozen(\u001b[39mdict\u001b[39;49m(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs))\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:534\u001b[0m, in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    532\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_variables\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[1;32m    533\u001b[0m     \u001b[39mreturn\u001b[39;00m FrozenDict(\n\u001b[0;32m--> 534\u001b[0m         (k, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mopen_store_variable(k, v)) \u001b[39mfor\u001b[39;00m k, v \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mzarr_group\u001b[39m.\u001b[39marrays()\n\u001b[1;32m    535\u001b[0m     )\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:510\u001b[0m, in \u001b[0;36mZarrStore.open_store_variable\u001b[0;34m(self, name, zarr_array)\u001b[0m\n\u001b[1;32m    508\u001b[0m data \u001b[39m=\u001b[39m indexing\u001b[39m.\u001b[39mLazilyIndexedArray(ZarrArrayWrapper(zarr_array))\n\u001b[1;32m    509\u001b[0m try_nczarr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_mode \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m--> 510\u001b[0m dimensions, attributes \u001b[39m=\u001b[39m _get_zarr_dims_and_attrs(\n\u001b[1;32m    511\u001b[0m     zarr_array, DIMENSION_KEY, try_nczarr\n\u001b[1;32m    512\u001b[0m )\n\u001b[1;32m    513\u001b[0m attributes \u001b[39m=\u001b[39m \u001b[39mdict\u001b[39m(attributes)\n\u001b[1;32m    515\u001b[0m \u001b[39m# TODO: this should not be needed once\u001b[39;00m\n\u001b[1;32m    516\u001b[0m \u001b[39m# https://github.com/zarr-developers/zarr-python/issues/1269 is resolved.\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/site-packages/xarray/backends/zarr.py:220\u001b[0m, in \u001b[0;36m_get_zarr_dims_and_attrs\u001b[0;34m(zarr_obj, dimension_key, try_nczarr)\u001b[0m\n\u001b[1;32m    218\u001b[0m \u001b[39m# NCZarr defines dimensions through metadata in .zarray\u001b[39;00m\n\u001b[1;32m    219\u001b[0m zarray_path \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mjoin(zarr_obj\u001b[39m.\u001b[39mpath, \u001b[39m\"\u001b[39m\u001b[39m.zarray\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m--> 220\u001b[0m zarray \u001b[39m=\u001b[39m json\u001b[39m.\u001b[39;49mloads(zarr_obj\u001b[39m.\u001b[39;49mstore[zarray_path])\n\u001b[1;32m    221\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    222\u001b[0m     \u001b[39m# NCZarr uses Fully Qualified Names\u001b[39;00m\n\u001b[1;32m    223\u001b[0m     dimensions \u001b[39m=\u001b[39m [\n\u001b[1;32m    224\u001b[0m         os\u001b[39m.\u001b[39mpath\u001b[39m.\u001b[39mbasename(dim) \u001b[39mfor\u001b[39;00m dim \u001b[39min\u001b[39;00m zarray[\u001b[39m\"\u001b[39m\u001b[39m_NCZARR_ARRAY\u001b[39m\u001b[39m\"\u001b[39m][\u001b[39m\"\u001b[39m\u001b[39mdimrefs\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[1;32m    225\u001b[0m     ]\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/ml4gland/lib/python3.9/json/__init__.py:339\u001b[0m, in \u001b[0;36mloads\u001b[0;34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[0m\n\u001b[1;32m    337\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    338\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(s, (\u001b[39mbytes\u001b[39m, \u001b[39mbytearray\u001b[39m)):\n\u001b[0;32m--> 339\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mthe JSON object must be str, bytes or bytearray, \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    340\u001b[0m                         \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mnot \u001b[39m\u001b[39m{\u001b[39;00ms\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m    341\u001b[0m     s \u001b[39m=\u001b[39m s\u001b[39m.\u001b[39mdecode(detect_encoding(s), \u001b[39m'\u001b[39m\u001b[39msurrogatepass\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m    343\u001b[0m \u001b[39mif\u001b[39;00m (\u001b[39mcls\u001b[39m \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m object_hook \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    344\u001b[0m         parse_int \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m parse_float \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m\n\u001b[1;32m    345\u001b[0m         parse_constant \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m object_pairs_hook \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m kw):\n",
      "\u001b[0;31mTypeError\u001b[0m: the JSON object must be str, bytes or bytearray, not dict"
     ]
    }
   ],
   "source": [
    "sd.open_zarr(\"/cellar/users/aklie/data/datasets/pbmc3k/pbmc3k.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9 ml4gland",
   "language": "python",
   "name": "ml4gland"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
