{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The history saving thread hit an unexpected error (DatabaseError('database disk image is malformed')).History will not be written to the database.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import zarr\n",
    "from pathlib import Path\n",
    "from seqdata import Table, FlatFASTA, GenomeFASTA, BigWig, BAM\n",
    "import seqdata as sd\n",
    "import xarray as xr\n",
    "import pysam\n",
    "import random\n",
    "import pickle\n",
    "import pyBigWig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a temporary directory for the output\n",
    "os.makedirs(Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'tmp', exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get infiles\n",
    "variable_fasta_in = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'variable.fa'\n",
    "variable_bed_in = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'variable.bed'\n",
    "fixed_bed_in = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'fixed.bed'\n",
    "out = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'tmp'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `GenomeFasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data with Python as true representation\n",
    "def read_fasta(file_path):\n",
    "    sequences = {}\n",
    "    with open(file_path, 'r') as file:\n",
    "        sequence_id = None\n",
    "        sequence_lines = []\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if sequence_id:\n",
    "                    sequences[sequence_id] = ''.join(sequence_lines)\n",
    "                sequence_id = line[1:]  # Remove the '>'\n",
    "                sequence_lines = []\n",
    "            else:\n",
    "                sequence_lines.append(line)\n",
    "        if sequence_id:\n",
    "            sequences[sequence_id] = ''.join(sequence_lines)\n",
    "    return sequences\n",
    "\n",
    "\n",
    "def test_GenomeFASTA(input_fasta, input_bed, temp_dir, batch_size=50, fixed_length=20, length_dim=\"_length\"):\n",
    "    \"\"\"\n",
    "    Tests the GenomeFASTA class for a single input FASTA file and BED coordinates.\n",
    "\n",
    "    Parameters:\n",
    "        input_fasta (str or Path): Path to the input FASTA file.\n",
    "        bed (list): List of tuples (chromosome, start, end) representing BED coordinates.\n",
    "        temp_dir (str or Path): Directory where temporary files will be stored.\n",
    "        batch_size (int): Batch size for the GenomeFASTA reader.\n",
    "        fixed_length (int): Fixed length for the sequences when writing to Zarr.\n",
    "        length_dim (str): Name of the length dimension in the Zarr output.\n",
    "\n",
    "    Raises:\n",
    "        AssertionError: If any test fails.\n",
    "    \"\"\"\n",
    "    temp_dir = Path(temp_dir)\n",
    "    temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "    fasta_out = temp_dir / 'output.fasta.zarr'\n",
    "\n",
    "    bed = pd.read_csv(input_bed, sep=\"\\t\", header=None)\n",
    "    fasta_sequences = read_fasta(input_fasta)\n",
    "    if fixed_length:\n",
    "        midpoint = (bed[1] + bed[2]) // 2\n",
    "        bed[1] = midpoint - fixed_length // 2\n",
    "        bed[2] = midpoint + fixed_length // 2\n",
    "    true = [fasta_sequences[chrom][start:end] for chrom, start, end in bed.values]\n",
    "    bed[\"strand\"] = \"+\"\n",
    "\n",
    "    # Test instantiation of the GenomeFASTA reader class\n",
    "    genomefasta_reader = GenomeFASTA(\n",
    "        name=\"seq\",\n",
    "        fasta=input_fasta,\n",
    "        batch_size=batch_size,\n",
    "    )\n",
    "    assert isinstance(genomefasta_reader, GenomeFASTA), \"GenomeFASTA reader instantiation failed.\"\n",
    "\n",
    "    # Read in the data using the reader\n",
    "    iterator = genomefasta_reader._reader(\n",
    "        bed=bed,\n",
    "        f=pysam.FastaFile(input_fasta)\n",
    "    )\n",
    "    _read = [seq.decode('utf-8') for seq in iterator]\n",
    "    assert np.array_equal(_read, true), \"GenomeFASTA reader failed to read in the correct values.\"\n",
    "\n",
    "    # Test writing to Zarr\n",
    "    genomefasta_reader._write(\n",
    "        fasta_out,\n",
    "        bed=bed,\n",
    "        fixed_length=fixed_length,\n",
    "        sequence_dim=\"_sequence\",\n",
    "        length_dim=length_dim,\n",
    "        overwrite=True\n",
    "    )\n",
    "    zarr.consolidate_metadata(fasta_out)\n",
    "\n",
    "    # Test round-trip reading\n",
    "    data = sd.open_zarr(fasta_out)\n",
    "\n",
    "    # Verify length and sequence dimensions\n",
    "    if fixed_length:\n",
    "        assert data[\"seq\"].shape[1] == fixed_length, \"Length dimension is incorrect.\"\n",
    "    assert data[\"seq\"].shape[0] == len(bed), \"Sequence dimension is incorrect.\"\n",
    "\n",
    "    # Verify that the data is the same\n",
    "    if fixed_length:\n",
    "        seqs = [''.join(row.astype(str)) for row in data[\"seq\"].values]\n",
    "    else:\n",
    "        seqs = data[\"seq\"].values.astype(str)\n",
    "    print(seqs)\n",
    "    print(true)\n",
    "    assert np.array_equal(true, seqs), \"Sequences do not match.\"\n",
    "\n",
    "    # Clean up temporary files\n",
    "    for f in temp_dir.iterdir():\n",
    "        if f.is_dir():\n",
    "            os.system(f'rm -r {f}')\n",
    "        else:\n",
    "            f.unlink()\n",
    "\n",
    "    print(\"All tests passed successfully!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'iter_rows'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3303169/1061945308.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Test the GenomeFASTA class on variable length sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtest_GenomeFASTA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvariable_fasta_in\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvariable_bed_in\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfixed_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlength_dim\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"_length\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/tmp/ipykernel_3303169/2244675442.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(input_fasta, input_bed, temp_dir, batch_size, fixed_length, length_dim)\u001b[0m\n\u001b[1;32m     58\u001b[0m     iterator = genomefasta_reader._reader(\n\u001b[1;32m     59\u001b[0m         \u001b[0mbed\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbed\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m         \u001b[0mf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpysam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFastaFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_fasta\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m     )\n\u001b[0;32m---> 62\u001b[0;31m     \u001b[0m_read\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m     \u001b[0;32massert\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray_equal\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_read\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"GenomeFASTA reader failed to read in the correct values.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m     \u001b[0;31m# Test writing to Zarr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_3303169/2244675442.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(.0)\u001b[0m\n\u001b[0;32m---> 62\u001b[0;31m \u001b[0;32mdef\u001b[0m \u001b[0mtest_GenomeFASTA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_fasta\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minput_bed\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtemp_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfixed_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlength_dim\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"_length\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m     \"\"\"\n\u001b[1;32m     64\u001b[0m     \u001b[0mTests\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mGenomeFASTA\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0msingle\u001b[0m \u001b[0minput\u001b[0m \u001b[0mFASTA\u001b[0m \u001b[0mfile\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mBED\u001b[0m \u001b[0mcoordinates\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/projects/ML4GLand/SeqData/seqdata/_io/readers/fasta.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, bed, f)\u001b[0m\n\u001b[1;32m    160\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_reader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbed\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mpysam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFastaFile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 161\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtqdm\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miter_rows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtotal\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbed\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    162\u001b[0m             \u001b[0mcontig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mTuple\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    163\u001b[0m             \u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcontig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ascii\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    164\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mpad_len\u001b[0m \u001b[0;34m:=\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5898\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5899\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5900\u001b[0m         ):\n\u001b[1;32m   5901\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5902\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'iter_rows'"
     ]
    }
   ],
   "source": [
    "# Test the GenomeFASTA class on variable length sequences\n",
    "test_GenomeFASTA(variable_fasta_in, variable_bed_in, out, batch_size=50, fixed_length=False, length_dim=\"_length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 65682.61it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 50231.19it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATCATCACAACTTGGCCTAA', 'ACATAGTAGGTAACACAACC', 'AGCTCCTTAGcCTGTcAgGG', 'TatAGCCtAAGCcTAACTTC', 'TTTTACACAGCCGCAAGTCT', 'CACCCGTTAAATTTGCCGAG', 'TGGCAACCCGTGTCATGAGG', 'TTATGTAGGCGGGATTCTGA', 'TAAGCGGCCGGTTATGACTA', 'TAGTGAGTAAGAACGATCGC', 'TTTCGACTCGAGATTTATGG', 'CCGAGCTCAGAATACTAAAA', 'TAAGCACAGCATTCTACACA', 'GAGGTAAACCTGACATAGTA']\n",
      "['ATCATCACAACTTGGCCTAA', 'ACATAGTAGGTAACACAACC', 'AGCTCCTTAGcCTGTcAgGG', 'TatAGCCtAAGCcTAACTTC', 'TTTTACACAGCCGCAAGTCT', 'CACCCGTTAAATTTGCCGAG', 'TGGCAACCCGTGTCATGAGG', 'TTATGTAGGCGGGATTCTGA', 'TAAGCGGCCGGTTATGACTA', 'TAGTGAGTAAGAACGATCGC', 'TTTCGACTCGAGATTTATGG', 'CCGAGCTCAGAATACTAAAA', 'TAAGCACAGCATTCTACACA', 'GAGGTAAACCTGACATAGTA']\n",
      "All tests passed successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Test the GenomeFASTA class on variable length sequences\n",
    "test_GenomeFASTA(variable_fasta_in, variable_bed_in, out, batch_size=50, fixed_length=20, length_dim=\"_length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 72137.91it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 49594.81it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CGATCATCACAACTTGGCCT', 'GTTAACATAGTAGGTAACAC', 'AACAGCTCCTTAGcCTGTcA', 'TTatAGCCtAAGCcTAACTT', 'TATTTTACACAGCCGCAAGT', 'TACTCACCCGTTAAATTTGC', 'GTTGGCAACCCGTGTCATGA', 'GTTTATGTAGGCGGGATTCT', 'ATAAGCGGCCGGTTATGACT', 'CATAGTGAGTAAGAACGATC', 'ATGGCTTTCGACTCGAGATT', 'CACCGAGCTCAGAATACTAA', 'GTAAGCACAGCATTCTACAC', 'GAGGTAAACCTGACATAGTA']\n",
      "['CGATCATCACAACTTGGCCT', 'GTTAACATAGTAGGTAACAC', 'AACAGCTCCTTAGcCTGTcA', 'TTatAGCCtAAGCcTAACTT', 'TATTTTACACAGCCGCAAGT', 'TACTCACCCGTTAAATTTGC', 'GTTGGCAACCCGTGTCATGA', 'GTTTATGTAGGCGGGATTCT', 'ATAAGCGGCCGGTTATGACT', 'CATAGTGAGTAAGAACGATC', 'ATGGCTTTCGACTCGAGATT', 'CACCGAGCTCAGAATACTAA', 'GTAAGCACAGCATTCTACAC', 'GAGGTAAACCTGACATAGTA']\n",
      "All tests passed successfully!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# Test the GenomeFASTA class on fixed length sequences\n",
    "test_GenomeFASTA(variable_fasta_in, fixed_bed_in, out, batch_size=50, fixed_length=20, length_dim=\"_length\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `BAM`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[PosixPath('/cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated1.bam'),\n",
       " PosixPath('/cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated2.bam'),\n",
       " PosixPath('/cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated3.bam'),\n",
       " PosixPath('/cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated4.bam'),\n",
       " PosixPath('/cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated5.bam')]"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list of bams with simulated{1-5}.bam\n",
    "bams = [Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / f'simulated{i}.bam' for i in range(1, 6)]\n",
    "bams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in coverage\n",
    "path_coverage = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'variable.bedcov.pkl'\n",
    "coverages = pickle.load(open(path_coverage, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>47</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>46</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>174</td>\n",
       "      <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>18</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr3</td>\n",
       "      <td>78</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr4</td>\n",
       "      <td>35</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr4</td>\n",
       "      <td>87</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr5</td>\n",
       "      <td>40</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr5</td>\n",
       "      <td>156</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr6</td>\n",
       "      <td>19</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr6</td>\n",
       "      <td>61</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr7</td>\n",
       "      <td>12</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr7</td>\n",
       "      <td>153</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0    1    2\n",
       "0   chr1    4   24\n",
       "1   chr1   47   67\n",
       "2   chr2   46   66\n",
       "3   chr2  174  194\n",
       "4   chr3   18   38\n",
       "5   chr3   78   98\n",
       "6   chr4   35   55\n",
       "7   chr4   87  107\n",
       "8   chr5   40   60\n",
       "9   chr5  156  176\n",
       "10  chr6   19   39\n",
       "11  chr6   61   81\n",
       "12  chr7   12   32\n",
       "13  chr7  153  173"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the bed file\n",
    "bed = pd.read_csv(fixed_bed_in, sep=\"\\t\", header=None)\n",
    "bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region: chr1:4-28 -> chr1:4-24\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 2 2 3]\n",
      "Offsets: 0, -4\n",
      "Region: chr1:47-76 -> chr1:47-67\n",
      "[3 3 3 2 2 1 2 1 1 1 1 1 2 2 2 2 1 1 1 1 1 2 1 1 1 1 1 1 1]\n",
      "Offsets: 0, -9\n",
      "Region: chr2:46-72 -> chr2:46-66\n",
      "[2 1 1 2 2 2 2 2 2 2 2 3 3 2 4 4 3 3 5 5 6 6 6 6 4 4]\n",
      "Offsets: 0, -6\n",
      "Region: chr2:174-197 -> chr2:174-194\n",
      "[2 1 1 1 2 3 3 4 4 4 5 6 6 6 5 4 3 2 3 4 3 2 3]\n",
      "Offsets: 0, -3\n",
      "Region: chr3:18-43 -> chr3:18-38\n",
      "[1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0]\n",
      "Offsets: 0, -5\n",
      "Region: chr3:78-106 -> chr3:78-98\n",
      "[1 0 0 1 3 4 5 5 5 5 5 5 5 4 2 1 0 0 1 3 3 3 3 3 3 3 3 3]\n",
      "Offsets: 0, -8\n",
      "Region: chr4:35-60 -> chr4:35-55\n",
      "[1 0 0 0 2 2 3 3 4 4 4 4 4 4 2 2 2 2 1 4 4 4 4 5 5]\n",
      "Offsets: 0, -5\n",
      "Region: chr4:87-111 -> chr4:87-107\n",
      "[2 3 4 4 4 4 3 2 3 3 3 4 4 4 4 4 5 7 6 6 6 4 5 6]\n",
      "Offsets: 0, -4\n",
      "Region: chr5:40-62 -> chr5:40-60\n",
      "[1 1 0 0 0 0 0 1 2 2 2 2 2 2 2 3 3 2 1 1 1 2]\n",
      "Offsets: 0, -2\n",
      "Region: chr5:156-181 -> chr5:156-176\n",
      "[1 2 2 2 2 2 2 2 2 2 3 2 2 4 4 4 6 6 6 6 6 6 6 4 4]\n",
      "Offsets: 0, -5\n",
      "Region: chr6:19-49 -> chr6:19-39\n",
      "[3 2 2 2 2 2 2 2 2 2 1 2 2 3 3 5 4 4 4 4 4 4 5 5 5 3 3 3 4 4]\n",
      "Offsets: 0, -10\n",
      "Region: chr6:61-85 -> chr6:61-81\n",
      "[6 6 5 5 5 5 4 5 5 6 4 4 4 4 3 4 4 3 3 3 3 3 3 3]\n",
      "Offsets: 0, -4\n",
      "Region: chr7:12-34 -> chr7:12-32\n",
      "[0 0 0 0 1 1 1 1 1 1 1 1 2 2 1 2 2 2 2 3 3 3]\n",
      "Offsets: 0, -2\n",
      "Region: chr7:153-174 -> chr7:153-173\n",
      "[2 2 2 2 2 3 3 4 4 4 4 4 5 5 6 5 5 5 5 6 7]\n",
      "Offsets: 0, -1\n",
      "Region: chr1:4-28 -> chr1:4-24\n",
      "[1 1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0]\n",
      "Offsets: 0, -4\n",
      "Region: chr1:47-76 -> chr1:47-67\n",
      "[1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 2 2 2 2 2 2 2 2 1 1 0 0 0 1]\n",
      "Offsets: 0, -9\n",
      "Region: chr2:46-72 -> chr2:46-66\n",
      "[2 2 1 2 2 2 3 4 4 4 3 3 3 2 2 3 2 2 3 3 3 4 4 4 4 3]\n",
      "Offsets: 0, -6\n",
      "Region: chr2:174-197 -> chr2:174-194\n",
      "[2 1 0 0 0 0 1 3 3 3 3 3 3 4 4 4 3 1 2 2 2 2 5]\n",
      "Offsets: 0, -3\n",
      "Region: chr3:18-43 -> chr3:18-38\n",
      "[1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 3 3 4 4 5 4 4 4 5 4]\n",
      "Offsets: 0, -5\n",
      "Region: chr3:78-106 -> chr3:78-98\n",
      "[2 4 4 4 5 6 7 7 6 6 6 3 3 4 3 2 4 4 4 5 6 6 6 5 5 5 2 2]\n",
      "Offsets: 0, -8\n",
      "Region: chr4:35-60 -> chr4:35-55\n",
      "[0 1 1 1 1 1 1 1 1 1 1 0 1 3 3 3 5 5 6 6 6 6 6 5 5]\n",
      "Offsets: 0, -5\n",
      "Region: chr4:87-111 -> chr4:87-107\n",
      "[1 1 1 1 1 2 2 2 2 2 1 1 1 1 1 2 2 2 3 4 5 5 5 6]\n",
      "Offsets: 0, -4\n",
      "Region: chr5:40-62 -> chr5:40-60\n",
      "[0 0 0 0 0 0 0 1 1 1 1 1 3 3 3 3 3 2 2 4 4 5]\n",
      "Offsets: 0, -2\n",
      "Region: chr5:156-181 -> chr5:156-176\n",
      "[1 1 1 1 2 2 2 3 3 5 5 5 5 5 4 5 5 4 4 4 4 5 6 6 8]\n",
      "Offsets: 0, -5\n",
      "Region: chr6:19-49 -> chr6:19-39\n",
      "[0 0 0 1 1 1 1 1 1 2 2 2 2 1 1 1 2 2 3 2 2 3 4 5 6 7 6 7 6 7]\n",
      "Offsets: 0, -10\n",
      "Region: chr6:61-85 -> chr6:61-81\n",
      "[ 9  9 11  9  7  7  5  6  7  8  7  9  7  8  8  7  7  7  5  4  4  2  3  2]\n",
      "Offsets: 0, -4\n",
      "Region: chr7:12-34 -> chr7:12-32\n",
      "[5 5 4 3 4 5 4 6 7 7 7 8 8 8 7 6 6 5 4 5 6 5]\n",
      "Offsets: 0, -2\n",
      "Region: chr7:153-174 -> chr7:153-173\n",
      "[3 3 3 3 3 3 4 3 3 3 4 3 4 5 5 4 4 4 4 4 4]\n",
      "Offsets: 0, -1\n",
      "Region: chr1:4-28 -> chr1:4-24\n",
      "[1 1 1 1 1 1 1 2 2 2 1 1 1 2 3 4 4 3 3 3 3 3 4 3]\n",
      "Offsets: 0, -4\n",
      "Region: chr1:47-76 -> chr1:47-67\n",
      "[1 1 2 2 2 3 4 4 4 5 5 5 4 4 3 2 1 2 2 2 4 5 5 5 5 6 6 6 7]\n",
      "Offsets: 0, -9\n",
      "Region: chr2:46-72 -> chr2:46-66\n",
      "[2 2 2 2 3 3 3 3 3 1 1 1 1 1 2 3 3 3 3 4 5 5 5 6 5 5]\n",
      "Offsets: 0, -6\n",
      "Region: chr2:174-197 -> chr2:174-194\n",
      "[4 3 2 2 1 2 3 5 5 7 6 6 6 6 6 6 5 3 3 1 1 1 2]\n",
      "Offsets: 0, -3\n",
      "Region: chr3:18-43 -> chr3:18-38\n",
      "[3 3 2 2 2 2 1 2 1 1 1 2 2 4 5 6 6 6 7 7 7 6 6 4 3]\n",
      "Offsets: 0, -5\n",
      "Region: chr3:78-106 -> chr3:78-98\n",
      "[0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0 0 1 1 1 2 2 2 2 2 2 2]\n",
      "Offsets: 0, -8\n",
      "Region: chr4:35-60 -> chr4:35-55\n",
      "[0 0 0 0 0 0 0 0 0 1 1 1 2 2 3 3 5 5 5 4 4 4 4 4 5]\n",
      "Offsets: 0, -5\n",
      "Region: chr4:87-111 -> chr4:87-107\n",
      "[3 2 2 3 4 4 4 6 6 6 6 8 8 7 6 7 6 4 5 6 5 3 5 5]\n",
      "Offsets: 0, -4\n",
      "Region: chr5:40-62 -> chr5:40-60\n",
      "[3 3 3 3 2 1 1 0 0 1 2 2 3 3 3 3 3 3 3 2 3 3]\n",
      "Offsets: 0, -2\n",
      "Region: chr5:156-181 -> chr5:156-176\n",
      "[3 2 2 2 1 1 2 2 3 3 3 3 3 3 4 4 4 5 4 3 3 3 3 4 3]\n",
      "Offsets: 0, -5\n",
      "Region: chr6:19-49 -> chr6:19-39\n",
      "[3 3 1 1 1 1 1 2 2 2 2 2 3 3 4 4 4 2 2 2 2 2 1 1 0 0 0 1 2 4]\n",
      "Offsets: 0, -10\n",
      "Region: chr6:61-85 -> chr6:61-81\n",
      "[2 3 3 3 3 5 5 7 7 8 7 6 5 5 5 3 3 1 1 0 1 1 2 2]\n",
      "Offsets: 0, -4\n",
      "Region: chr7:12-34 -> chr7:12-32\n",
      "[2 2 1 1 1 1 0 1 1 1 1 2 2 2 2 2 2 1 1 1 1 0]\n",
      "Offsets: 0, -2\n",
      "Region: chr7:153-174 -> chr7:153-173\n",
      "[1 1 1 1 1 1 2 2 3 2 2 2 2 2 3 3 3 3 2 2 2]\n",
      "Offsets: 0, -1\n",
      "Region: chr1:4-28 -> chr1:4-24\n",
      "[1 1 1 1 1 2 2 2 3 2 2 2 2 2 3 2 2 2 1 2 3 3 3 4]\n",
      "Offsets: 0, -4\n",
      "Region: chr1:47-76 -> chr1:47-67\n",
      "[6 4 4 3 2 1 2 2 4 4 5 5 5 5 6 7 6 6 4 3 2 2 2 3 2 2 2 2 2]\n",
      "Offsets: 0, -9\n",
      "Region: chr2:46-72 -> chr2:46-66\n",
      "[0 0 0 0 0 0 0 1 1 1 1 1 1 2 2 2 2 1 1 1 1 1 2 2 3 3]\n",
      "Offsets: 0, -6\n",
      "Region: chr2:174-197 -> chr2:174-194\n",
      "[1 1 1 2 2 2 2 2 2 2 2 3 3 2 2 2 1 1 2 2 2 2 2]\n",
      "Offsets: 0, -3\n",
      "Region: chr3:18-43 -> chr3:18-38\n",
      "[0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1]\n",
      "Offsets: 0, -5\n",
      "Region: chr3:78-106 -> chr3:78-98\n",
      "[1 1 2 2 2 2 2 2 2 2 1 1 0 0 0 1 1 2 2 2 2 2 2 2 2 1 1 0]\n",
      "Offsets: 0, -8\n",
      "Region: chr4:35-60 -> chr4:35-55\n",
      "[0 0 0 0 0 0 0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0]\n",
      "Offsets: 0, -5\n",
      "Region: chr4:87-111 -> chr4:87-107\n",
      "[1 2 2 2 2 4 4 5 6 7 8 7 7 7 7 4 5 4 3 2 4 4 5 6]\n",
      "Offsets: 0, -4\n",
      "Region: chr5:40-62 -> chr5:40-60\n",
      "[1 1 2 2 3 3 3 3 4 5 4 4 3 3 2 3 4 7 7 7 7 7]\n",
      "Offsets: 0, -2\n",
      "Region: chr5:156-181 -> chr5:156-176\n",
      "[3 3 1 1 1 0 1 3 3 3 4 4 4 4 4 4 3 1 1 2 1 2 2 2 2]\n",
      "Offsets: 0, -5\n",
      "Region: chr6:19-49 -> chr6:19-39\n",
      "[1 1 1 1 1 1 1 1 1 0 0 1 1 1 2 2 2 2 2 2 3 2 2 3 3 3 4 4 4 4]\n",
      "Offsets: 0, -10\n",
      "Region: chr6:61-85 -> chr6:61-81\n",
      "[4 4 5 4 3 3 2 1 1 2 2 2 1 1 1 1 1 2 2 1 1 1 1 1]\n",
      "Offsets: 0, -4\n",
      "Region: chr7:12-34 -> chr7:12-32\n",
      "[0 1 2 2 2 2 2 2 2 2 3 2 2 2 2 2 3 5 5 5 4 4]\n",
      "Offsets: 0, -2\n",
      "Region: chr7:153-174 -> chr7:153-173\n",
      "[3 3 3 2 3 4 3 3 4 4 4 5 6 6 5 4 4 3 2 3 4]\n",
      "Offsets: 0, -1\n",
      "Region: chr1:4-28 -> chr1:4-24\n",
      "[1 1 1 1 2 2 1 2 2 2 2 3 4 4 3 3 4 3 4 7 7 7 8 8]\n",
      "Offsets: 0, -4\n",
      "Region: chr1:47-76 -> chr1:47-67\n",
      "[7 5 5 3 2 4 4 3 4 4 3 3 4 5 6 5 5 5 4 4 7 7 6 5 4 3 3 3 4]\n",
      "Offsets: 0, -9\n",
      "Region: chr2:46-72 -> chr2:46-66\n",
      "[1 2 2 3 3 5 4 5 5 5 5 5 5 4 5 3 5 4 5 5 6 5 6 6 5 5]\n",
      "Offsets: 0, -6\n",
      "Region: chr2:174-197 -> chr2:174-194\n",
      "[2 2 2 2 1 1 1 1 0 1 1 1 1 2 2 2 2 2 3 2 2 2 2]\n",
      "Offsets: 0, -3\n",
      "Region: chr3:18-43 -> chr3:18-38\n",
      "[1 1 1 1 1 3 3 3 2 2 2 2 2 3 3 1 1 1 1 1 1 1 1 0 0]\n",
      "Offsets: 0, -5\n",
      "Region: chr3:78-106 -> chr3:78-98\n",
      "[2 2 2 3 3 3 3 3 3 3 1 1 1 0 0 2 2 2 3 3 3 3 3 3 3 1 1 1]\n",
      "Offsets: 0, -8\n",
      "Region: chr4:35-60 -> chr4:35-55\n",
      "[0 1 3 4 4 4 4 4 4 4 4 3 1 0 1 1 2 4 5 5 5 5 5 5 5]\n",
      "Offsets: 0, -5\n",
      "Region: chr4:87-111 -> chr4:87-107\n",
      "[3 2 2 2 2 2 3 3 3 2 1 1 1 2 3 4 4 4 4 4 4 4 4 3]\n",
      "Offsets: 0, -4\n",
      "Region: chr5:40-62 -> chr5:40-60\n",
      "[1 1 1 1 0 1 1 1 1 3 3 5 5 5 5 4 4 4 4 2 2 0]\n",
      "Offsets: 0, -2\n",
      "Region: chr5:156-181 -> chr5:156-176\n",
      "[0 0 0 0 0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0 0 1]\n",
      "Offsets: 0, -5\n",
      "Region: chr6:19-49 -> chr6:19-39\n",
      "[0 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 2 2 2 4 4 4 4 4 5 4 4 4 4]\n",
      "Offsets: 0, -10\n",
      "Region: chr6:61-85 -> chr6:61-81\n",
      "[5 4 4 0 0 0 2 2 3 3 3 3 3 3 3 3 1 1 0 0 0 1 1 1]\n",
      "Offsets: 0, -4\n",
      "Region: chr7:12-34 -> chr7:12-32\n",
      "[1 0 0 0 0 1 2 2 2 2 2 2 2 3 4 4 4 4 4 4 4 4]\n",
      "Offsets: 0, -2\n",
      "Region: chr7:153-174 -> chr7:153-173\n",
      "[2 3 3 3 3 4 4 4 4 4 4 3 5 5 5 4 5 5 5 6 7]\n",
      "Offsets: 0, -1\n"
     ]
    }
   ],
   "source": [
    "# Subset the coverage to the bed file coordinates\n",
    "new_cov = {}\n",
    "# for each bam file\n",
    "for bam in bams:\n",
    "    bam_name = bam.name\n",
    "    new_cov[bam_name] = {}\n",
    "    for i, (region, coverage) in enumerate(coverages[bam_name].items()):\n",
    "        coverage_interval = region.split(\":\")[1]\n",
    "        coverage_start, coverage_end = map(int, coverage_interval.split(\"-\"))\n",
    "        start_offset = coverage_start - bed[1].values[i]\n",
    "        end_offset = bed[2].values[i] - coverage_end\n",
    "        new_region = f\"{bed[0].values[i]}:{bed[1].values[i]}-{bed[2].values[i]}\"\n",
    "        print(f\"Region: {region} -> {new_region}\")\n",
    "        print(coverage)\n",
    "        print(f\"Offsets: {start_offset}, {end_offset}\")\n",
    "        if end_offset == 0:\n",
    "            new_cov[bam_name][new_region] = coverage[start_offset:]\n",
    "        else:\n",
    "            new_cov[bam_name][new_region] = coverage[start_offset:end_offset]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_BAM(bam_files, bed_file, coverages, temp_dir, batch_size=50, sequence_dim=\"_sequence\"):\n",
    "    \"\"\"\n",
    "    Tests the BAM class for one or more input BAM files and their coverage.\n",
    "\n",
    "    Parameters:\n",
    "        bam_files (list): List of paths to BAM files.\n",
    "        bed_file (str or Path): Path to the BED file.\n",
    "        coverage_file (str or Path): Path to the pickled coverage file.\n",
    "        temp_dir (str or Path): Directory where temporary files will be stored.\n",
    "        batch_size (int): Batch size for the BAM reader.\n",
    "        sequence_dim (str): Name of the sequence dimension in the Zarr output.\n",
    "\n",
    "    Raises:\n",
    "        AssertionError: If any test fails.\n",
    "    \"\"\"\n",
    "    temp_dir = Path(temp_dir)\n",
    "    temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "    bam_out = temp_dir / 'output.bam.zarr'\n",
    "\n",
    "    # Test instantiation of the BAM reader class\n",
    "    bam_reader = BAM(\n",
    "        name=\"cov\",\n",
    "        bams=bam_files,\n",
    "        samples=[bam.name for bam in bam_files],\n",
    "        batch_size=batch_size,\n",
    "    )\n",
    "    assert isinstance(bam_reader, BAM), \"BAM reader instantiation failed.\"\n",
    "\n",
    "    # Read in the BED file\n",
    "    bed = pd.read_csv(bed_file, sep=\"\\t\", header=None)\n",
    "\n",
    "    # Verify each BAM file individually\n",
    "    for bam_file in bam_files:\n",
    "        print(f\"Testing {bam_file}\")\n",
    "        bam_name = bam_file.name\n",
    "        iterator = bam_reader._reader(bed=bed, f=pysam.AlignmentFile(bam_file))\n",
    "        _read = [list(seq) for seq in iterator]\n",
    "\n",
    "        # Verify that the data matches the expected coverage\n",
    "        for i, (region, coverage) in enumerate(coverages[bam_name].items()):\n",
    "            print(f\"Region {region}: {list(_read[i])}, {coverage}\")\n",
    "            assert np.array_equal(coverage, _read[i]), f\"Region {region} in {bam_name} does not match.\"\n",
    "\n",
    "    # Test writing to Zarr\n",
    "    bed[\"strand\"] = \"+\"\n",
    "    bam_reader._write(\n",
    "        bam_out,\n",
    "        bed=bed,\n",
    "        fixed_length=False,\n",
    "        sequence_dim=sequence_dim,\n",
    "        overwrite=True\n",
    "    )\n",
    "    zarr.consolidate_metadata(bam_out)\n",
    "\n",
    "    # Test round-trip reading\n",
    "    data = sd.open_zarr(bam_out)\n",
    "    return data\n",
    "    # Verify that the data matches the expected coverage for all regions\n",
    "    for file_index, bam_file in enumerate(bam_files):\n",
    "        print(f\"Testing {bam_file}\")\n",
    "        bam_name = bam_file.name\n",
    "        for region_index, (region, coverage) in enumerate(coverages[bam_name].items()):\n",
    "            print(data['cov'][region_index].shape)\n",
    "            #print(f\"Region {region}: {list(data['cov'].values[:, i])}, {coverage}\")\n",
    "            assert np.array_equal(coverage, data[\"cov\"].values[region_index, file_index]), f\"Region {region} in {bam_name} does not match.\"\n",
    "\n",
    "    # Clean up temporary files\n",
    "    for f in temp_dir.iterdir():\n",
    "        if f.is_dir():\n",
    "            os.system(f'rm -r {f}')\n",
    "        else:\n",
    "            f.unlink()\n",
    "\n",
    "    print(\"All tests passed successfully!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing /cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated1.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 12576.62it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3], [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 2 2 3]\n",
      "Region chr1:47-76: [3, 3, 3, 2, 2, 1, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1], [3 3 3 2 2 1 2 1 1 1 1 1 2 2 2 2 1 1 1 1 1 2 1 1 1 1 1 1 1]\n",
      "Region chr2:46-72: [2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 4, 4, 3, 3, 5, 5, 6, 6, 6, 6, 4, 4], [2 1 1 2 2 2 2 2 2 2 2 3 3 2 4 4 3 3 5 5 6 6 6 6 4 4]\n",
      "Region chr2:174-197: [2, 1, 1, 1, 2, 3, 3, 4, 4, 4, 5, 6, 6, 6, 5, 4, 3, 2, 3, 4, 3, 2, 3], [2 1 1 1 2 3 3 4 4 4 5 6 6 6 5 4 3 2 3 4 3 2 3]\n",
      "Region chr3:18-43: [1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0], [1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0]\n",
      "Region chr3:78-106: [1, 0, 0, 1, 3, 4, 5, 5, 5, 5, 5, 5, 5, 4, 2, 1, 0, 0, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3], [1 0 0 1 3 4 5 5 5 5 5 5 5 4 2 1 0 0 1 3 3 3 3 3 3 3 3 3]\n",
      "Region chr4:35-60: [1, 0, 0, 0, 2, 2, 3, 3, 4, 4, 4, 4, 4, 4, 2, 2, 2, 2, 1, 4, 4, 4, 4, 5, 5], [1 0 0 0 2 2 3 3 4 4 4 4 4 4 2 2 2 2 1 4 4 4 4 5 5]\n",
      "Region chr4:87-111: [2, 3, 4, 4, 4, 4, 3, 2, 3, 3, 3, 4, 4, 4, 4, 4, 5, 7, 6, 6, 6, 4, 5, 6], [2 3 4 4 4 4 3 2 3 3 3 4 4 4 4 4 5 7 6 6 6 4 5 6]\n",
      "Region chr5:40-62: [1, 1, 0, 0, 0, 0, 0, 1, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 1, 1, 1, 2], [1 1 0 0 0 0 0 1 2 2 2 2 2 2 2 3 3 2 1 1 1 2]\n",
      "Region chr5:156-181: [1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 4, 4, 4, 6, 6, 6, 6, 6, 6, 6, 4, 4], [1 2 2 2 2 2 2 2 2 2 3 2 2 4 4 4 6 6 6 6 6 6 6 4 4]\n",
      "Region chr6:19-49: [3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 3, 3, 5, 4, 4, 4, 4, 4, 4, 5, 5, 5, 3, 3, 3, 4, 4], [3 2 2 2 2 2 2 2 2 2 1 2 2 3 3 5 4 4 4 4 4 4 5 5 5 3 3 3 4 4]\n",
      "Region chr6:61-85: [6, 6, 5, 5, 5, 5, 4, 5, 5, 6, 4, 4, 4, 4, 3, 4, 4, 3, 3, 3, 3, 3, 3, 3], [6 6 5 5 5 5 4 5 5 6 4 4 4 4 3 4 4 3 3 3 3 3 3 3]\n",
      "Region chr7:12-34: [0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 1, 2, 2, 2, 2, 3, 3, 3], [0 0 0 0 1 1 1 1 1 1 1 1 2 2 1 2 2 2 2 3 3 3]\n",
      "Region chr7:153-174: [2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4, 4, 5, 5, 6, 5, 5, 5, 5, 6, 7], [2 2 2 2 2 3 3 4 4 4 4 4 5 5 6 5 5 5 5 6 7]\n",
      "Testing /cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated2.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 12175.05it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0], [1 1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0]\n",
      "Region chr1:47-76: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 1], [1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 2 2 2 2 2 2 2 2 1 1 0 0 0 1]\n",
      "Region chr2:46-72: [2, 2, 1, 2, 2, 2, 3, 4, 4, 4, 3, 3, 3, 2, 2, 3, 2, 2, 3, 3, 3, 4, 4, 4, 4, 3], [2 2 1 2 2 2 3 4 4 4 3 3 3 2 2 3 2 2 3 3 3 4 4 4 4 3]\n",
      "Region chr2:174-197: [2, 1, 0, 0, 0, 0, 1, 3, 3, 3, 3, 3, 3, 4, 4, 4, 3, 1, 2, 2, 2, 2, 5], [2 1 0 0 0 0 1 3 3 3 3 3 3 4 4 4 3 1 2 2 2 2 5]\n",
      "Region chr3:18-43: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 3, 3, 4, 4, 5, 4, 4, 4, 5, 4], [1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 3 3 4 4 5 4 4 4 5 4]\n",
      "Region chr3:78-106: [2, 4, 4, 4, 5, 6, 7, 7, 6, 6, 6, 3, 3, 4, 3, 2, 4, 4, 4, 5, 6, 6, 6, 5, 5, 5, 2, 2], [2 4 4 4 5 6 7 7 6 6 6 3 3 4 3 2 4 4 4 5 6 6 6 5 5 5 2 2]\n",
      "Region chr4:35-60: [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 3, 3, 3, 5, 5, 6, 6, 6, 6, 6, 5, 5], [0 1 1 1 1 1 1 1 1 1 1 0 1 3 3 3 5 5 6 6 6 6 6 5 5]\n",
      "Region chr4:87-111: [1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 2, 3, 4, 5, 5, 5, 6], [1 1 1 1 1 2 2 2 2 2 1 1 1 1 1 2 2 2 3 4 5 5 5 6]\n",
      "Region chr5:40-62: [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3, 2, 2, 4, 4, 5], [0 0 0 0 0 0 0 1 1 1 1 1 3 3 3 3 3 2 2 4 4 5]\n",
      "Region chr5:156-181: [1, 1, 1, 1, 2, 2, 2, 3, 3, 5, 5, 5, 5, 5, 4, 5, 5, 4, 4, 4, 4, 5, 6, 6, 8], [1 1 1 1 2 2 2 3 3 5 5 5 5 5 4 5 5 4 4 4 4 5 6 6 8]\n",
      "Region chr6:19-49: [0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 2, 2, 3, 2, 2, 3, 4, 5, 6, 7, 6, 7, 6, 7], [0 0 0 1 1 1 1 1 1 2 2 2 2 1 1 1 2 2 3 2 2 3 4 5 6 7 6 7 6 7]\n",
      "Region chr6:61-85: [9, 9, 11, 9, 7, 7, 5, 6, 7, 8, 7, 9, 7, 8, 8, 7, 7, 7, 5, 4, 4, 2, 3, 2], [ 9  9 11  9  7  7  5  6  7  8  7  9  7  8  8  7  7  7  5  4  4  2  3  2]\n",
      "Region chr7:12-34: [5, 5, 4, 3, 4, 5, 4, 6, 7, 7, 7, 8, 8, 8, 7, 6, 6, 5, 4, 5, 6, 5], [5 5 4 3 4 5 4 6 7 7 7 8 8 8 7 6 6 5 4 5 6 5]\n",
      "Region chr7:153-174: [3, 3, 3, 3, 3, 3, 4, 3, 3, 3, 4, 3, 4, 5, 5, 4, 4, 4, 4, 4, 4], [3 3 3 3 3 3 4 3 3 3 4 3 4 5 5 4 4 4 4 4 4]\n",
      "Testing /cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated3.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 11978.84it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 1, 1, 1, 2, 3, 4, 4, 3, 3, 3, 3, 3, 4, 3], [1 1 1 1 1 1 1 2 2 2 1 1 1 2 3 4 4 3 3 3 3 3 4 3]\n",
      "Region chr1:47-76: [1, 1, 2, 2, 2, 3, 4, 4, 4, 5, 5, 5, 4, 4, 3, 2, 1, 2, 2, 2, 4, 5, 5, 5, 5, 6, 6, 6, 7], [1 1 2 2 2 3 4 4 4 5 5 5 4 4 3 2 1 2 2 2 4 5 5 5 5 6 6 6 7]\n",
      "Region chr2:46-72: [2, 2, 2, 2, 3, 3, 3, 3, 3, 1, 1, 1, 1, 1, 2, 3, 3, 3, 3, 4, 5, 5, 5, 6, 5, 5], [2 2 2 2 3 3 3 3 3 1 1 1 1 1 2 3 3 3 3 4 5 5 5 6 5 5]\n",
      "Region chr2:174-197: [4, 3, 2, 2, 1, 2, 3, 5, 5, 7, 6, 6, 6, 6, 6, 6, 5, 3, 3, 1, 1, 1, 2], [4 3 2 2 1 2 3 5 5 7 6 6 6 6 6 6 5 3 3 1 1 1 2]\n",
      "Region chr3:18-43: [3, 3, 2, 2, 2, 2, 1, 2, 1, 1, 1, 2, 2, 4, 5, 6, 6, 6, 7, 7, 7, 6, 6, 4, 3], [3 3 2 2 2 2 1 2 1 1 1 2 2 4 5 6 6 6 7 7 7 6 6 4 3]\n",
      "Region chr3:78-106: [0, 0, 0, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 0, 0, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2], [0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0 0 1 1 1 2 2 2 2 2 2 2]\n",
      "Region chr4:35-60: [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 2, 2, 3, 3, 5, 5, 5, 4, 4, 4, 4, 4, 5], [0 0 0 0 0 0 0 0 0 1 1 1 2 2 3 3 5 5 5 4 4 4 4 4 5]\n",
      "Region chr4:87-111: [3, 2, 2, 3, 4, 4, 4, 6, 6, 6, 6, 8, 8, 7, 6, 7, 6, 4, 5, 6, 5, 3, 5, 5], [3 2 2 3 4 4 4 6 6 6 6 8 8 7 6 7 6 4 5 6 5 3 5 5]\n",
      "Region chr5:40-62: [3, 3, 3, 3, 2, 1, 1, 0, 0, 1, 2, 2, 3, 3, 3, 3, 3, 3, 3, 2, 3, 3], [3 3 3 3 2 1 1 0 0 1 2 2 3 3 3 3 3 3 3 2 3 3]\n",
      "Region chr5:156-181: [3, 2, 2, 2, 1, 1, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 5, 4, 3, 3, 3, 3, 4, 3], [3 2 2 2 1 1 2 2 3 3 3 3 3 3 4 4 4 5 4 3 3 3 3 4 3]\n",
      "Region chr6:19-49: [3, 3, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 1, 2, 4], [3 3 1 1 1 1 1 2 2 2 2 2 3 3 4 4 4 2 2 2 2 2 1 1 0 0 0 1 2 4]\n",
      "Region chr6:61-85: [2, 3, 3, 3, 3, 5, 5, 7, 7, 8, 7, 6, 5, 5, 5, 3, 3, 1, 1, 0, 1, 1, 2, 2], [2 3 3 3 3 5 5 7 7 8 7 6 5 5 5 3 3 1 1 0 1 1 2 2]\n",
      "Region chr7:12-34: [2, 2, 1, 1, 1, 1, 0, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 0], [2 2 1 1 1 1 0 1 1 1 1 2 2 2 2 2 2 1 1 1 1 0]\n",
      "Region chr7:153-174: [1, 1, 1, 1, 1, 1, 2, 2, 3, 2, 2, 2, 2, 2, 3, 3, 3, 3, 2, 2, 2], [1 1 1 1 1 1 2 2 3 2 2 2 2 2 3 3 3 3 2 2 2]\n",
      "Testing /cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated4.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 13446.36it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1, 1, 1, 1, 1, 2, 2, 2, 3, 2, 2, 2, 2, 2, 3, 2, 2, 2, 1, 2, 3, 3, 3, 4], [1 1 1 1 1 2 2 2 3 2 2 2 2 2 3 2 2 2 1 2 3 3 3 4]\n",
      "Region chr1:47-76: [6, 4, 4, 3, 2, 1, 2, 2, 4, 4, 5, 5, 5, 5, 6, 7, 6, 6, 4, 3, 2, 2, 2, 3, 2, 2, 2, 2, 2], [6 4 4 3 2 1 2 2 4 4 5 5 5 5 6 7 6 6 4 3 2 2 2 3 2 2 2 2 2]\n",
      "Region chr2:46-72: [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 3, 3], [0 0 0 0 0 0 0 1 1 1 1 1 1 2 2 2 2 1 1 1 1 1 2 2 3 3]\n",
      "Region chr2:174-197: [1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2], [1 1 1 2 2 2 2 2 2 2 2 3 3 2 2 2 1 1 2 2 2 2 2]\n",
      "Region chr3:18-43: [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1], [0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1]\n",
      "Region chr3:78-106: [1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 0, 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 0], [1 1 2 2 2 2 2 2 2 2 1 1 0 0 0 1 1 2 2 2 2 2 2 2 2 1 1 0]\n",
      "Region chr4:35-60: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 0], [0 0 0 0 0 0 0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0]\n",
      "Region chr4:87-111: [1, 2, 2, 2, 2, 4, 4, 5, 6, 7, 8, 7, 7, 7, 7, 4, 5, 4, 3, 2, 4, 4, 5, 6], [1 2 2 2 2 4 4 5 6 7 8 7 7 7 7 4 5 4 3 2 4 4 5 6]\n",
      "Region chr5:40-62: [1, 1, 2, 2, 3, 3, 3, 3, 4, 5, 4, 4, 3, 3, 2, 3, 4, 7, 7, 7, 7, 7], [1 1 2 2 3 3 3 3 4 5 4 4 3 3 2 3 4 7 7 7 7 7]\n",
      "Region chr5:156-181: [3, 3, 1, 1, 1, 0, 1, 3, 3, 3, 4, 4, 4, 4, 4, 4, 3, 1, 1, 2, 1, 2, 2, 2, 2], [3 3 1 1 1 0 1 3 3 3 4 4 4 4 4 4 3 1 1 2 1 2 2 2 2]\n",
      "Region chr6:19-49: [1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 2, 2, 2, 2, 2, 2, 3, 2, 2, 3, 3, 3, 4, 4, 4, 4], [1 1 1 1 1 1 1 1 1 0 0 1 1 1 2 2 2 2 2 2 3 2 2 3 3 3 4 4 4 4]\n",
      "Region chr6:61-85: [4, 4, 5, 4, 3, 3, 2, 1, 1, 2, 2, 2, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1], [4 4 5 4 3 3 2 1 1 2 2 2 1 1 1 1 1 2 2 1 1 1 1 1]\n",
      "Region chr7:12-34: [0, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2, 2, 3, 5, 5, 5, 4, 4], [0 1 2 2 2 2 2 2 2 2 3 2 2 2 2 2 3 5 5 5 4 4]\n",
      "Region chr7:153-174: [3, 3, 3, 2, 3, 4, 3, 3, 4, 4, 4, 5, 6, 6, 5, 4, 4, 3, 2, 3, 4], [3 3 3 2 3 4 3 3 4 4 4 5 6 6 5 4 4 3 2 3 4]\n",
      "Testing /cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated5.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 7639.90it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1, 1, 1, 1, 2, 2, 1, 2, 2, 2, 2, 3, 4, 4, 3, 3, 4, 3, 4, 7, 7, 7, 8, 8], [1 1 1 1 2 2 1 2 2 2 2 3 4 4 3 3 4 3 4 7 7 7 8 8]\n",
      "Region chr1:47-76: [7, 5, 5, 3, 2, 4, 4, 3, 4, 4, 3, 3, 4, 5, 6, 5, 5, 5, 4, 4, 7, 7, 6, 5, 4, 3, 3, 3, 4], [7 5 5 3 2 4 4 3 4 4 3 3 4 5 6 5 5 5 4 4 7 7 6 5 4 3 3 3 4]\n",
      "Region chr2:46-72: [1, 2, 2, 3, 3, 5, 4, 5, 5, 5, 5, 5, 5, 4, 5, 3, 5, 4, 5, 5, 6, 5, 6, 6, 5, 5], [1 2 2 3 3 5 4 5 5 5 5 5 5 4 5 3 5 4 5 5 6 5 6 6 5 5]\n",
      "Region chr2:174-197: [2, 2, 2, 2, 1, 1, 1, 1, 0, 1, 1, 1, 1, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2], [2 2 2 2 1 1 1 1 0 1 1 1 1 2 2 2 2 2 3 2 2 2 2]\n",
      "Region chr3:18-43: [1, 1, 1, 1, 1, 3, 3, 3, 2, 2, 2, 2, 2, 3, 3, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], [1 1 1 1 1 3 3 3 2 2 2 2 2 3 3 1 1 1 1 1 1 1 1 0 0]\n",
      "Region chr3:78-106: [2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 1, 1, 1, 0, 0, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 1, 1, 1], [2 2 2 3 3 3 3 3 3 3 1 1 1 0 0 2 2 2 3 3 3 3 3 3 3 1 1 1]\n",
      "Region chr4:35-60: [0, 1, 3, 4, 4, 4, 4, 4, 4, 4, 4, 3, 1, 0, 1, 1, 2, 4, 5, 5, 5, 5, 5, 5, 5], [0 1 3 4 4 4 4 4 4 4 4 3 1 0 1 1 2 4 5 5 5 5 5 5 5]\n",
      "Region chr4:87-111: [3, 2, 2, 2, 2, 2, 3, 3, 3, 2, 1, 1, 1, 2, 3, 4, 4, 4, 4, 4, 4, 4, 4, 3], [3 2 2 2 2 2 3 3 3 2 1 1 1 2 3 4 4 4 4 4 4 4 4 3]\n",
      "Region chr5:40-62: [1, 1, 1, 1, 0, 1, 1, 1, 1, 3, 3, 5, 5, 5, 5, 4, 4, 4, 4, 2, 2, 0], [1 1 1 1 0 1 1 1 1 3 3 5 5 5 5 4 4 4 4 2 2 0]\n",
      "Region chr5:156-181: [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 0, 0, 1], [0 0 0 0 0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0 0 1]\n",
      "Region chr6:19-49: [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 4, 4, 4, 4, 4, 5, 4, 4, 4, 4], [0 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 2 2 2 4 4 4 4 4 5 4 4 4 4]\n",
      "Region chr6:61-85: [5, 4, 4, 0, 0, 0, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 1, 1, 0, 0, 0, 1, 1, 1], [5 4 4 0 0 0 2 2 3 3 3 3 3 3 3 3 1 1 0 0 0 1 1 1]\n",
      "Region chr7:12-34: [1, 0, 0, 0, 0, 1, 2, 2, 2, 2, 2, 2, 2, 3, 4, 4, 4, 4, 4, 4, 4, 4], [1 0 0 0 0 1 2 2 2 2 2 2 2 3 4 4 4 4 4 4 4 4]\n",
      "Region chr7:153-174: [2, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 3, 5, 5, 5, 4, 5, 5, 5, 6, 7], [2 3 3 3 3 4 4 4 4 4 4 3 5 5 5 4 5 5 5 6 7]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 7486.96it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 8631.52it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 13333.39it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 16686.63it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 13394.22it/s]\n"
     ]
    }
   ],
   "source": [
    "data = test_BAM(bams, variable_bed_in, coverages, out, batch_size=50, sequence_dim=\"_sequence\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(coverages[\"simulated1.bam\"].values())[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data.sel(cov_sample=\"simulated1.bam\").cov.values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing /cellar/users/aklie/projects/ML4GLand/SeqData/tests/data/simulated1.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 15208.56it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-24: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1], [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1]\n",
      "Region chr1:47-67: [3, 3, 3, 2, 2, 1, 2, 1, 1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1], [3 3 3 2 2 1 2 1 1 1 1 1 2 2 2 2 1 1 1 1]\n",
      "Region chr2:46-66: [2, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 4, 4, 3, 3, 5, 5], [2 1 1 2 2 2 2 2 2 2 2 3 3 2 4 4 3 3 5 5]\n",
      "Region chr2:174-194: [2, 1, 1, 1, 2, 3, 3, 4, 4, 4, 5, 6, 6, 6, 5, 4, 3, 2, 3, 4], [2 1 1 1 2 3 3 4 4 4 5 6 6 6 5 4 3 2 3 4]\n",
      "Region chr3:18-38: [1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0], [1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0]\n",
      "Region chr3:78-98: [1, 0, 0, 1, 3, 4, 5, 5, 5, 5, 5, 5, 5, 4, 2, 1, 0, 0, 1, 3], [1 0 0 1 3 4 5 5 5 5 5 5 5 4 2 1 0 0 1 3]\n",
      "Region chr4:35-55: [1, 0, 0, 0, 2, 2, 3, 3, 4, 4, 4, 4, 4, 4, 2, 2, 2, 2, 1, 4], [1 0 0 0 2 2 3 3 4 4 4 4 4 4 2 2 2 2 1 4]\n",
      "Region chr4:87-107: [2, 3, 4, 4, 4, 4, 3, 2, 3, 3, 3, 4, 4, 4, 4, 4, 5, 7, 6, 6], [2 3 4 4 4 4 3 2 3 3 3 4 4 4 4 4 5 7 6 6]\n",
      "Region chr5:40-60: [1, 1, 0, 0, 0, 0, 0, 1, 2, 2, 2, 2, 2, 2, 2, 3, 3, 2, 1, 1], [1 1 0 0 0 0 0 1 2 2 2 2 2 2 2 3 3 2 1 1]\n",
      "Region chr5:156-176: [1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 4, 4, 4, 6, 6, 6, 6], [1 2 2 2 2 2 2 2 2 2 3 2 2 4 4 4 6 6 6 6]\n",
      "Region chr6:19-39: [3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 3, 3, 5, 4, 4, 4, 4], [3 2 2 2 2 2 2 2 2 2 1 2 2 3 3 5 4 4 4 4]\n",
      "Region chr6:61-81: [6, 6, 5, 5, 5, 5, 4, 5, 5, 6, 4, 4, 4, 4, 3, 4, 4, 3, 3, 3], [6 6 5 5 5 5 4 5 5 6 4 4 4 4 3 4 4 3 3 3]\n",
      "Region chr7:12-32: [0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 1, 2, 2, 2, 2, 3], [0 0 0 0 1 1 1 1 1 1 1 1 2 2 1 2 2 2 2 3]\n",
      "Region chr7:153-173: [2, 2, 2, 2, 2, 3, 3, 4, 4, 4, 4, 4, 5, 5, 6, 5, 5, 5, 5, 6], [2 2 2 2 2 3 3 4 4 4 4 4 5 5 6 5 5 5 5 6]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 12804.24it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "html[data-theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 0 20px 0 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: inline-block;\n",
       "  opacity: 0;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:focus + label {\n",
       "  border: 2px solid var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt; Size: 128B\n",
       "Dimensions:          (_sequence: 14, cov_sample: 1)\n",
       "Coordinates:\n",
       "  * cov_sample       (cov_sample) object 8B &#x27;simulated1.bam&#x27;\n",
       "Dimensions without coordinates: _sequence\n",
       "Data variables:\n",
       "    cov              (_sequence, cov_sample) object 112B dask.array&lt;chunksize=(14, 1), meta=np.ndarray&gt;\n",
       "    total_reads_cov  (cov_sample) uint64 8B dask.array&lt;chunksize=(1,), meta=np.ndarray&gt;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-ea708ea6-515b-4f10-8ab3-9fd6e064dd72' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ea708ea6-515b-4f10-8ab3-9fd6e064dd72' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>_sequence</span>: 14</li><li><span class='xr-has-index'>cov_sample</span>: 1</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-1d42b1b7-95c1-4f2e-9524-1310c287c24c' class='xr-section-summary-in' type='checkbox'  checked><label for='section-1d42b1b7-95c1-4f2e-9524-1310c287c24c' class='xr-section-summary' >Coordinates: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cov_sample</span></div><div class='xr-var-dims'>(cov_sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;simulated1.bam&#x27;</div><input id='attrs-9ec08bd2-5736-48df-bfed-d39b10bc35be' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-9ec08bd2-5736-48df-bfed-d39b10bc35be' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-bef8c3b1-cbed-4e9a-a5e3-f378c24b9600' class='xr-var-data-in' type='checkbox'><label for='data-bef8c3b1-cbed-4e9a-a5e3-f378c24b9600' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><pre>array([&#x27;simulated1.bam&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-c13eb0e7-3fa0-4181-a1ae-133ccb35ca3e' class='xr-section-summary-in' type='checkbox'  checked><label for='section-c13eb0e7-3fa0-4181-a1ae-133ccb35ca3e' class='xr-section-summary' >Data variables: <span>(2)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>cov</span></div><div class='xr-var-dims'>(_sequence, cov_sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(14, 1), meta=np.ndarray&gt;</div><input id='attrs-395423e8-dfc9-4cc9-8260-57c3e48935b5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-395423e8-dfc9-4cc9-8260-57c3e48935b5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-51c2027b-b0de-4156-b6f5-e9c6eca0f8b6' class='xr-var-data-in' type='checkbox'><label for='data-51c2027b-b0de-4156-b6f5-e9c6eca0f8b6' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 112 B </td>\n",
       "                        <td> 112 B </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (14, 1) </td>\n",
       "                        <td> (14, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 1 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> object numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"86\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"36\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"36\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"36\" y1=\"0\" x2=\"36\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 36.659534236895894,0.0 36.659534236895894,120.0 0.0,120.0\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"18.329767\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"56.659534\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,56.659534,60.000000)\">14</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li><li class='xr-var-item'><div class='xr-var-name'><span>total_reads_cov</span></div><div class='xr-var-dims'>(cov_sample)</div><div class='xr-var-dtype'>uint64</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(1,), meta=np.ndarray&gt;</div><input id='attrs-835b0b17-2ff6-4bdd-b8a1-09d733a6905d' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-835b0b17-2ff6-4bdd-b8a1-09d733a6905d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-9a0d910d-a89a-4a0c-bf24-92452e49f435' class='xr-var-data-in' type='checkbox'><label for='data-9a0d910d-a89a-4a0c-bf24-92452e49f435' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 8 B </td>\n",
       "                        <td> 8 B </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (1,) </td>\n",
       "                        <td> (1,) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 1 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> uint64 numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"170\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"120\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"120\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"120\" y1=\"0\" x2=\"120\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 120.0,0.0 120.0,120.0 0.0,120.0\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"60.000000\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >1</text>\n",
       "  <text x=\"140.000000\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,140.000000,60.000000)\">1</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-d2f154b0-3202-4b0c-8f13-276286388ff5' class='xr-section-summary-in' type='checkbox'  ><label for='section-d2f154b0-3202-4b0c-8f13-276286388ff5' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>cov_sample</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-4d2564fc-4265-4f30-a68e-bf89d40f3963' class='xr-index-data-in' type='checkbox'/><label for='index-4d2564fc-4265-4f30-a68e-bf89d40f3963' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;simulated1.bam&#x27;], dtype=&#x27;object&#x27;, name=&#x27;cov_sample&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-01e7c5e4-e0b3-4991-9c4d-520e5e67121b' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-01e7c5e4-e0b3-4991-9c4d-520e5e67121b' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset> Size: 128B\n",
       "Dimensions:          (_sequence: 14, cov_sample: 1)\n",
       "Coordinates:\n",
       "  * cov_sample       (cov_sample) object 8B 'simulated1.bam'\n",
       "Dimensions without coordinates: _sequence\n",
       "Data variables:\n",
       "    cov              (_sequence, cov_sample) object 112B dask.array<chunksize=(14, 1), meta=np.ndarray>\n",
       "    total_reads_cov  (cov_sample) uint64 8B dask.array<chunksize=(1,), meta=np.ndarray>"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_BAM([bams[0]], fixed_bed_in, {\"simulated1.bam\": new_cov[\"simulated1.bam\"]}, out, batch_size=50, sequence_dim=\"_sequence\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `BigWig`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_BigWig(bigwig_files, bed_file, chromsizes_file, coverages, temp_dir, batch_size=50, sequence_dim=\"_sequence\", fixed_length=False, length_dim=None):\n",
    "    \"\"\"\n",
    "    Tests the BigWig class for one or more input BigWig files.\n",
    "\n",
    "    Parameters:\n",
    "        bigwig_files (list): List of paths to BigWig files.\n",
    "        bed_file (str or Path): Path to the BED file.\n",
    "        chromsizes_file (str or Path): Path to the chromsizes file.\n",
    "        coverage_file (str or Path): Path to the pickled coverage file.\n",
    "        temp_dir (str or Path): Directory where temporary files will be stored.\n",
    "        batch_size (int): Batch size for the BigWig reader.\n",
    "        sequence_dim (str): Name of the sequence dimension in the Zarr output.\n",
    "\n",
    "    Raises:\n",
    "        AssertionError: If any test fails.\n",
    "    \"\"\"\n",
    "    temp_dir = Path(temp_dir)\n",
    "    temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "    bigwig_out = temp_dir / 'output.bigwig.zarr'\n",
    "\n",
    "    # Load chromsizes\n",
    "    chromsizes = {}\n",
    "    with open(chromsizes_file) as f:\n",
    "        for line in f:\n",
    "            chrom, size = line.strip().split()\n",
    "            chromsizes[chrom] = int(size)\n",
    "\n",
    "    # Test instantiation of the BigWig reader class\n",
    "    bigwig_reader = BigWig(\n",
    "        name=\"cov\",\n",
    "        bigwigs=bigwig_files,\n",
    "        samples=[bigwig.name for bigwig in bigwig_files],\n",
    "        batch_size=batch_size,\n",
    "    )\n",
    "    assert isinstance(bigwig_reader, BigWig), \"BigWig reader instantiation failed.\"\n",
    "\n",
    "    # Read in the BED file\n",
    "    bed = pd.read_csv(bed_file, sep=\"\\t\", header=None)\n",
    "\n",
    "    # Verify each BigWig file individually\n",
    "    for bigwig_file in bigwig_files:\n",
    "        bigwig_name = bigwig_file.name\n",
    "        iterator = bigwig_reader._reader(\n",
    "            bed=bed,\n",
    "            f=pyBigWig.open(str(bigwig_file)),\n",
    "            contig_lengths=chromsizes\n",
    "        )\n",
    "        _read = [list(seq) for seq in iterator]\n",
    "\n",
    "        # Verify that the data matches the expected coverage\n",
    "        for i, (region, coverage) in enumerate(coverages[bigwig_name].items()):\n",
    "            print(f\"Region {region}: {list(_read[i])}, {coverage}\")\n",
    "            assert np.array_equal(coverage, _read[i]), f\"Region {region} in {bigwig_name} does not match.\"\n",
    "\n",
    "        # Close the BigWig file\n",
    "        iterator.close()\n",
    "\n",
    "    # Test writing to Zarr\n",
    "    bed[\"strand\"] = \"+\"\n",
    "    bigwig_reader._write(\n",
    "        bigwig_out,\n",
    "        bed=bed,\n",
    "        fixed_length=fixed_length,\n",
    "        sequence_dim=sequence_dim,\n",
    "        length_dim=length_dim,\n",
    "        overwrite=True\n",
    "    )\n",
    "    zarr.consolidate_metadata(bigwig_out)\n",
    "\n",
    "    # Test round-trip reading\n",
    "    data = sd.open_zarr(bigwig_out)\n",
    "    return data\n",
    "    \n",
    "    # Verify that the data matches the expected coverage for all regions\n",
    "    for file_index, bigwig_file in enumerate(bigwig_files):\n",
    "        print(f\"Testing {bigwig_file}\")\n",
    "        bigwig_name = bigwig_file.name\n",
    "        for region_index, (region, coverage) in enumerate(coverages[bigwig_name].items()):\n",
    "            print(data['cov'][region_index].shape)\n",
    "            assert np.array_equal(coverage, data[\"cov\"].values[region_index, file_index]), f\"Region {region} in {bigwig_name} does not match.\"\n",
    "\n",
    "    # Clean up temporary files\n",
    "    for f in temp_dir.iterdir():\n",
    "        if f.is_dir():\n",
    "            os.system(f'rm -r {f}')\n",
    "        else:\n",
    "            f.unlink()\n",
    "\n",
    "    print(\"All tests passed successfully!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of bigwigs with simulated{1-5}.bw\n",
    "bigwigs = [Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / f'simulated{i}.bw' for i in range(1, 6)]\n",
    "chromsizes = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'variable.chrom.sizes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "# modify new cov to have .bw instead of .bam in keys\n",
    "new_cov_bw = {}\n",
    "for key, value in new_cov.items():\n",
    "    new_cov_bw[key.replace(\".bam\", \".bw\")] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "coverages_bw = {}\n",
    "for key, value in coverages.items():\n",
    "    coverages_bw[key.replace(\".bam\", \".bw\")] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 19560.38it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 3.0], [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 2 2 3]\n",
      "Region chr1:47-76: [3.0, 3.0, 3.0, 2.0, 2.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], [3 3 3 2 2 1 2 1 1 1 1 1 2 2 2 2 1 1 1 1 1 2 1 1 1 1 1 1 1]\n",
      "Region chr2:46-72: [2.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 2.0, 4.0, 4.0, 3.0, 3.0, 5.0, 5.0, 6.0, 6.0, 6.0, 6.0, 4.0, 4.0], [2 1 1 2 2 2 2 2 2 2 2 3 3 2 4 4 3 3 5 5 6 6 6 6 4 4]\n",
      "Region chr2:174-197: [2.0, 1.0, 1.0, 1.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 5.0, 6.0, 6.0, 6.0, 5.0, 4.0, 3.0, 2.0, 3.0, 4.0, 3.0, 2.0, 3.0], [2 1 1 1 2 3 3 4 4 4 5 6 6 6 5 4 3 2 3 4 3 2 3]\n",
      "Region chr3:18-43: [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], [1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0]\n",
      "Region chr3:78-106: [1.0, 0.0, 0.0, 1.0, 3.0, 4.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.0, 2.0, 1.0, 0.0, 0.0, 1.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0], [1 0 0 1 3 4 5 5 5 5 5 5 5 4 2 1 0 0 1 3 3 3 3 3 3 3 3 3]\n",
      "Region chr4:35-60: [1.0, 0.0, 0.0, 0.0, 2.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 2.0, 2.0, 2.0, 2.0, 1.0, 4.0, 4.0, 4.0, 4.0, 5.0, 5.0], [1 0 0 0 2 2 3 3 4 4 4 4 4 4 2 2 2 2 1 4 4 4 4 5 5]\n",
      "Region chr4:87-111: [2.0, 3.0, 4.0, 4.0, 4.0, 4.0, 3.0, 2.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 7.0, 6.0, 6.0, 6.0, 4.0, 5.0, 6.0], [2 3 4 4 4 4 3 2 3 3 3 4 4 4 4 4 5 7 6 6 6 4 5 6]\n",
      "Region chr5:40-62: [1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 2.0, 1.0, 1.0, 1.0, 2.0], [1 1 0 0 0 0 0 1 2 2 2 2 2 2 2 3 3 2 1 1 1 2]\n",
      "Region chr5:156-181: [1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 4.0, 4.0, 4.0, 6.0, 6.0, 6.0, 6.0, 6.0, 6.0, 6.0, 4.0, 4.0], [1 2 2 2 2 2 2 2 2 2 3 2 2 4 4 4 6 6 6 6 6 6 6 4 4]\n",
      "Region chr6:19-49: [3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 2.0, 2.0, 3.0, 3.0, 5.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 5.0, 5.0, 3.0, 3.0, 3.0, 4.0, 4.0], [3 2 2 2 2 2 2 2 2 2 1 2 2 3 3 5 4 4 4 4 4 4 5 5 5 3 3 3 4 4]\n",
      "Region chr6:61-85: [6.0, 6.0, 5.0, 5.0, 5.0, 5.0, 4.0, 5.0, 5.0, 6.0, 4.0, 4.0, 4.0, 4.0, 3.0, 4.0, 4.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0], [6 6 5 5 5 5 4 5 5 6 4 4 4 4 3 4 4 3 3 3 3 3 3 3]\n",
      "Region chr7:12-34: [0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0], [0 0 0 0 1 1 1 1 1 1 1 1 2 2 1 2 2 2 2 3 3 3]\n",
      "Region chr7:153-174: [2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 5.0, 6.0, 5.0, 5.0, 5.0, 5.0, 6.0, 7.0], [2 2 2 2 2 3 3 4 4 4 4 4 5 5 6 5 5 5 5 6 7]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 37568.94it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0], [1 1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0]\n",
      "Region chr1:47-76: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0], [1 1 1 1 1 1 1 1 1 1 0 0 0 1 1 2 2 2 2 2 2 2 2 1 1 0 0 0 1]\n",
      "Region chr2:46-72: [2.0, 2.0, 1.0, 2.0, 2.0, 2.0, 3.0, 4.0, 4.0, 4.0, 3.0, 3.0, 3.0, 2.0, 2.0, 3.0, 2.0, 2.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 3.0], [2 2 1 2 2 2 3 4 4 4 3 3 3 2 2 3 2 2 3 3 3 4 4 4 4 3]\n",
      "Region chr2:174-197: [2.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 3.0, 1.0, 2.0, 2.0, 2.0, 2.0, 5.0], [2 1 0 0 0 0 1 3 3 3 3 3 3 4 4 4 3 1 2 2 2 2 5]\n",
      "Region chr3:18-43: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 3.0, 3.0, 4.0, 4.0, 5.0, 4.0, 4.0, 4.0, 5.0, 4.0], [1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 3 3 4 4 5 4 4 4 5 4]\n",
      "Region chr3:78-106: [2.0, 4.0, 4.0, 4.0, 5.0, 6.0, 7.0, 7.0, 6.0, 6.0, 6.0, 3.0, 3.0, 4.0, 3.0, 2.0, 4.0, 4.0, 4.0, 5.0, 6.0, 6.0, 6.0, 5.0, 5.0, 5.0, 2.0, 2.0], [2 4 4 4 5 6 7 7 6 6 6 3 3 4 3 2 4 4 4 5 6 6 6 5 5 5 2 2]\n",
      "Region chr4:35-60: [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 3.0, 3.0, 3.0, 5.0, 5.0, 6.0, 6.0, 6.0, 6.0, 6.0, 5.0, 5.0], [0 1 1 1 1 1 1 1 1 1 1 0 1 3 3 3 5 5 6 6 6 6 6 5 5]\n",
      "Region chr4:87-111: [1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 3.0, 4.0, 5.0, 5.0, 5.0, 6.0], [1 1 1 1 1 2 2 2 2 2 1 1 1 1 1 2 2 2 3 4 5 5 5 6]\n",
      "Region chr5:40-62: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 3.0, 3.0, 3.0, 3.0, 3.0, 2.0, 2.0, 4.0, 4.0, 5.0], [0 0 0 0 0 0 0 1 1 1 1 1 3 3 3 3 3 2 2 4 4 5]\n",
      "Region chr5:156-181: [1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 3.0, 3.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.0, 5.0, 5.0, 4.0, 4.0, 4.0, 4.0, 5.0, 6.0, 6.0, 8.0], [1 1 1 1 2 2 2 3 3 5 5 5 5 5 4 5 5 4 4 4 4 5 6 6 8]\n",
      "Region chr6:19-49: [0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 2.0, 2.0, 3.0, 2.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 6.0, 7.0, 6.0, 7.0], [0 0 0 1 1 1 1 1 1 2 2 2 2 1 1 1 2 2 3 2 2 3 4 5 6 7 6 7 6 7]\n",
      "Region chr6:61-85: [9.0, 9.0, 11.0, 9.0, 7.0, 7.0, 5.0, 6.0, 7.0, 8.0, 7.0, 9.0, 7.0, 8.0, 8.0, 7.0, 7.0, 7.0, 5.0, 4.0, 4.0, 2.0, 3.0, 2.0], [ 9  9 11  9  7  7  5  6  7  8  7  9  7  8  8  7  7  7  5  4  4  2  3  2]\n",
      "Region chr7:12-34: [5.0, 5.0, 4.0, 3.0, 4.0, 5.0, 4.0, 6.0, 7.0, 7.0, 7.0, 8.0, 8.0, 8.0, 7.0, 6.0, 6.0, 5.0, 4.0, 5.0, 6.0, 5.0], [5 5 4 3 4 5 4 6 7 7 7 8 8 8 7 6 6 5 4 5 6 5]\n",
      "Region chr7:153-174: [3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 4.0, 3.0, 3.0, 3.0, 4.0, 3.0, 4.0, 5.0, 5.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0], [3 3 3 3 3 3 4 3 3 3 4 3 4 5 5 4 4 4 4 4 4]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 38631.75it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 2.0, 3.0, 4.0, 4.0, 3.0, 3.0, 3.0, 3.0, 3.0, 4.0, 3.0], [1 1 1 1 1 1 1 2 2 2 1 1 1 2 3 4 4 3 3 3 3 3 4 3]\n",
      "Region chr1:47-76: [1.0, 1.0, 2.0, 2.0, 2.0, 3.0, 4.0, 4.0, 4.0, 5.0, 5.0, 5.0, 4.0, 4.0, 3.0, 2.0, 1.0, 2.0, 2.0, 2.0, 4.0, 5.0, 5.0, 5.0, 5.0, 6.0, 6.0, 6.0, 7.0], [1 1 2 2 2 3 4 4 4 5 5 5 4 4 3 2 1 2 2 2 4 5 5 5 5 6 6 6 7]\n",
      "Region chr2:46-72: [2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 3.0, 3.0, 3.0, 3.0, 4.0, 5.0, 5.0, 5.0, 6.0, 5.0, 5.0], [2 2 2 2 3 3 3 3 3 1 1 1 1 1 2 3 3 3 3 4 5 5 5 6 5 5]\n",
      "Region chr2:174-197: [4.0, 3.0, 2.0, 2.0, 1.0, 2.0, 3.0, 5.0, 5.0, 7.0, 6.0, 6.0, 6.0, 6.0, 6.0, 6.0, 5.0, 3.0, 3.0, 1.0, 1.0, 1.0, 2.0], [4 3 2 2 1 2 3 5 5 7 6 6 6 6 6 6 5 3 3 1 1 1 2]\n",
      "Region chr3:18-43: [3.0, 3.0, 2.0, 2.0, 2.0, 2.0, 1.0, 2.0, 1.0, 1.0, 1.0, 2.0, 2.0, 4.0, 5.0, 6.0, 6.0, 6.0, 7.0, 7.0, 7.0, 6.0, 6.0, 4.0, 3.0], [3 3 2 2 2 2 1 2 1 1 1 2 2 4 5 6 6 6 7 7 7 6 6 4 3]\n",
      "Region chr3:78-106: [0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0], [0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0 0 1 1 1 2 2 2 2 2 2 2]\n",
      "Region chr4:35-60: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 3.0, 3.0, 5.0, 5.0, 5.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0], [0 0 0 0 0 0 0 0 0 1 1 1 2 2 3 3 5 5 5 4 4 4 4 4 5]\n",
      "Region chr4:87-111: [3.0, 2.0, 2.0, 3.0, 4.0, 4.0, 4.0, 6.0, 6.0, 6.0, 6.0, 8.0, 8.0, 7.0, 6.0, 7.0, 6.0, 4.0, 5.0, 6.0, 5.0, 3.0, 5.0, 5.0], [3 2 2 3 4 4 4 6 6 6 6 8 8 7 6 7 6 4 5 6 5 3 5 5]\n",
      "Region chr5:40-62: [3.0, 3.0, 3.0, 3.0, 2.0, 1.0, 1.0, 0.0, 0.0, 1.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 2.0, 3.0, 3.0], [3 3 3 3 2 1 1 0 0 1 2 2 3 3 3 3 3 3 3 2 3 3]\n",
      "Region chr5:156-181: [3.0, 2.0, 2.0, 2.0, 1.0, 1.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 5.0, 4.0, 3.0, 3.0, 3.0, 3.0, 4.0, 3.0], [3 2 2 2 1 1 2 2 3 3 3 3 3 3 4 4 4 5 4 3 3 3 3 4 3]\n",
      "Region chr6:19-49: [3.0, 3.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 2.0, 4.0], [3 3 1 1 1 1 1 2 2 2 2 2 3 3 4 4 4 2 2 2 2 2 1 1 0 0 0 1 2 4]\n",
      "Region chr6:61-85: [2.0, 3.0, 3.0, 3.0, 3.0, 5.0, 5.0, 7.0, 7.0, 8.0, 7.0, 6.0, 5.0, 5.0, 5.0, 3.0, 3.0, 1.0, 1.0, 0.0, 1.0, 1.0, 2.0, 2.0], [2 3 3 3 3 5 5 7 7 8 7 6 5 5 5 3 3 1 1 0 1 1 2 2]\n",
      "Region chr7:12-34: [2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 0.0], [2 2 1 1 1 1 0 1 1 1 1 2 2 2 2 2 2 1 1 1 1 0]\n",
      "Region chr7:153-174: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 2.0, 2.0, 2.0], [1 1 1 1 1 1 2 2 3 2 2 2 2 2 3 3 3 3 2 2 2]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 39146.84it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 1.0, 2.0, 3.0, 3.0, 3.0, 4.0], [1 1 1 1 1 2 2 2 3 2 2 2 2 2 3 2 2 2 1 2 3 3 3 4]\n",
      "Region chr1:47-76: [6.0, 4.0, 4.0, 3.0, 2.0, 1.0, 2.0, 2.0, 4.0, 4.0, 5.0, 5.0, 5.0, 5.0, 6.0, 7.0, 6.0, 6.0, 4.0, 3.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 2.0, 2.0], [6 4 4 3 2 1 2 2 4 4 5 5 5 5 6 7 6 6 4 3 2 2 2 3 2 2 2 2 2]\n",
      "Region chr2:46-72: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 3.0, 3.0], [0 0 0 0 0 0 0 1 1 1 1 1 1 2 2 2 2 1 1 1 1 1 2 2 3 3]\n",
      "Region chr2:174-197: [1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 2.0, 2.0, 2.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0], [1 1 1 2 2 2 2 2 2 2 2 3 3 2 2 2 1 1 2 2 2 2 2]\n",
      "Region chr3:18-43: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0], [0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 1 1 1 1]\n",
      "Region chr3:78-106: [1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 0.0], [1 1 2 2 2 2 2 2 2 2 1 1 0 0 0 1 1 2 2 2 2 2 2 2 2 1 1 0]\n",
      "Region chr4:35-60: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 0.0], [0 0 0 0 0 0 0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0]\n",
      "Region chr4:87-111: [1.0, 2.0, 2.0, 2.0, 2.0, 4.0, 4.0, 5.0, 6.0, 7.0, 8.0, 7.0, 7.0, 7.0, 7.0, 4.0, 5.0, 4.0, 3.0, 2.0, 4.0, 4.0, 5.0, 6.0], [1 2 2 2 2 4 4 5 6 7 8 7 7 7 7 4 5 4 3 2 4 4 5 6]\n",
      "Region chr5:40-62: [1.0, 1.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 4.0, 5.0, 4.0, 4.0, 3.0, 3.0, 2.0, 3.0, 4.0, 7.0, 7.0, 7.0, 7.0, 7.0], [1 1 2 2 3 3 3 3 4 5 4 4 3 3 2 3 4 7 7 7 7 7]\n",
      "Region chr5:156-181: [3.0, 3.0, 1.0, 1.0, 1.0, 0.0, 1.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 3.0, 1.0, 1.0, 2.0, 1.0, 2.0, 2.0, 2.0, 2.0], [3 3 1 1 1 0 1 3 3 3 4 4 4 4 4 4 3 1 1 2 1 2 2 2 2]\n",
      "Region chr6:19-49: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0], [1 1 1 1 1 1 1 1 1 0 0 1 1 1 2 2 2 2 2 2 3 2 2 3 3 3 4 4 4 4]\n",
      "Region chr6:61-85: [4.0, 4.0, 5.0, 4.0, 3.0, 3.0, 2.0, 1.0, 1.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0], [4 4 5 4 3 3 2 1 1 2 2 2 1 1 1 1 1 2 2 1 1 1 1 1]\n",
      "Region chr7:12-34: [0.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 5.0, 5.0, 5.0, 4.0, 4.0], [0 1 2 2 2 2 2 2 2 2 3 2 2 2 2 2 3 5 5 5 4 4]\n",
      "Region chr7:153-174: [3.0, 3.0, 3.0, 2.0, 3.0, 4.0, 3.0, 3.0, 4.0, 4.0, 4.0, 5.0, 6.0, 6.0, 5.0, 4.0, 4.0, 3.0, 2.0, 3.0, 4.0], [3 3 3 2 3 4 3 3 4 4 4 5 6 6 5 4 4 3 2 3 4]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 30631.33it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-28: [1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0, 4.0, 4.0, 3.0, 3.0, 4.0, 3.0, 4.0, 7.0, 7.0, 7.0, 8.0, 8.0], [1 1 1 1 2 2 1 2 2 2 2 3 4 4 3 3 4 3 4 7 7 7 8 8]\n",
      "Region chr1:47-76: [7.0, 5.0, 5.0, 3.0, 2.0, 4.0, 4.0, 3.0, 4.0, 4.0, 3.0, 3.0, 4.0, 5.0, 6.0, 5.0, 5.0, 5.0, 4.0, 4.0, 7.0, 7.0, 6.0, 5.0, 4.0, 3.0, 3.0, 3.0, 4.0], [7 5 5 3 2 4 4 3 4 4 3 3 4 5 6 5 5 5 4 4 7 7 6 5 4 3 3 3 4]\n",
      "Region chr2:46-72: [1.0, 2.0, 2.0, 3.0, 3.0, 5.0, 4.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.0, 5.0, 3.0, 5.0, 4.0, 5.0, 5.0, 6.0, 5.0, 6.0, 6.0, 5.0, 5.0], [1 2 2 3 3 5 4 5 5 5 5 5 5 4 5 3 5 4 5 5 6 5 6 6 5 5]\n",
      "Region chr2:174-197: [2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 2.0, 2.0], [2 2 2 2 1 1 1 1 0 1 1 1 1 2 2 2 2 2 3 2 2 2 2]\n",
      "Region chr3:18-43: [1.0, 1.0, 1.0, 1.0, 1.0, 3.0, 3.0, 3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0], [1 1 1 1 1 3 3 3 2 2 2 2 2 3 3 1 1 1 1 1 1 1 1 0 0]\n",
      "Region chr3:78-106: [2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 1.0, 1.0, 1.0, 0.0, 0.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 1.0, 1.0, 1.0], [2 2 2 3 3 3 3 3 3 3 1 1 1 0 0 2 2 2 3 3 3 3 3 3 3 1 1 1]\n",
      "Region chr4:35-60: [0.0, 1.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 3.0, 1.0, 0.0, 1.0, 1.0, 2.0, 4.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0], [0 1 3 4 4 4 4 4 4 4 4 3 1 0 1 1 2 4 5 5 5 5 5 5 5]\n",
      "Region chr4:87-111: [3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 2.0, 1.0, 1.0, 1.0, 2.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 3.0], [3 2 2 2 2 2 3 3 3 2 1 1 1 2 3 4 4 4 4 4 4 4 4 3]\n",
      "Region chr5:40-62: [1.0, 1.0, 1.0, 1.0, 0.0, 1.0, 1.0, 1.0, 1.0, 3.0, 3.0, 5.0, 5.0, 5.0, 5.0, 4.0, 4.0, 4.0, 4.0, 2.0, 2.0, 0.0], [1 1 1 1 0 1 1 1 1 3 3 5 5 5 5 4 4 4 4 2 2 0]\n",
      "Region chr5:156-181: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0], [0 0 0 0 0 0 0 0 0 1 1 1 2 2 2 2 2 2 2 1 1 1 0 0 1]\n",
      "Region chr6:19-49: [0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 4.0, 4.0, 4.0, 4.0], [0 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 2 2 2 2 4 4 4 4 4 5 4 4 4 4]\n",
      "Region chr6:61-85: [5.0, 4.0, 4.0, 0.0, 0.0, 0.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0], [5 4 4 0 0 0 2 2 3 3 3 3 3 3 3 3 1 1 0 0 0 1 1 1]\n",
      "Region chr7:12-34: [1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0], [1 0 0 0 0 1 2 2 2 2 2 2 2 3 4 4 4 4 4 4 4 4]\n",
      "Region chr7:153-174: [2.0, 3.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 3.0, 5.0, 5.0, 5.0, 4.0, 5.0, 5.0, 5.0, 6.0, 7.0], [2 3 3 3 3 4 4 4 4 4 4 3 5 5 5 4 5 5 5 6 7]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 29716.73it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 29716.73it/s]\n",
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 37617.08it/s]\n",
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 38404.35it/s]\n",
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 40302.17it/s]\n",
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 31621.03it/s]\n"
     ]
    }
   ],
   "source": [
    "# Test the BigWig class on variable length sequences\n",
    "data = test_BigWig(bigwigs, variable_bed_in, chromsizes, coverages_bw, out, batch_size=50, fixed_length=False, sequence_dim=\"_sequence\", length_dim=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "html[data-theme=dark],\n",
       "body[data-theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 0 20px 0 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: inline-block;\n",
       "  opacity: 0;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:focus + label {\n",
       "  border: 2px solid var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt; Size: 600B\n",
       "Dimensions:     (_sequence: 14, cov_sample: 5)\n",
       "Coordinates:\n",
       "  * cov_sample  (cov_sample) object 40B &#x27;simulated1.bw&#x27; ... &#x27;simulated5.bw&#x27;\n",
       "Dimensions without coordinates: _sequence\n",
       "Data variables:\n",
       "    cov         (_sequence, cov_sample) object 560B dask.array&lt;chunksize=(14, 1), meta=np.ndarray&gt;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-3a874fc1-fb54-41ba-9ef6-4e9100022c29' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-3a874fc1-fb54-41ba-9ef6-4e9100022c29' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span>_sequence</span>: 14</li><li><span class='xr-has-index'>cov_sample</span>: 5</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-73d24a37-fd78-4e87-aad3-8871066b9868' class='xr-section-summary-in' type='checkbox'  checked><label for='section-73d24a37-fd78-4e87-aad3-8871066b9868' class='xr-section-summary' >Coordinates: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>cov_sample</span></div><div class='xr-var-dims'>(cov_sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;simulated1.bw&#x27; ... &#x27;simulated5.bw&#x27;</div><input id='attrs-15d2d0d6-b350-4008-9109-ae20dc4ca62c' class='xr-var-attrs-in' type='checkbox' ><label for='attrs-15d2d0d6-b350-4008-9109-ae20dc4ca62c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-c44ede74-cb4b-43fd-8e15-81f9b07dbb3f' class='xr-var-data-in' type='checkbox'><label for='data-c44ede74-cb4b-43fd-8e15-81f9b07dbb3f' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'><dt><span>_FillValue :</span></dt><dd>0</dd></dl></div><div class='xr-var-data'><pre>array([&#x27;simulated1.bw&#x27;, &#x27;simulated2.bw&#x27;, &#x27;simulated3.bw&#x27;, &#x27;simulated4.bw&#x27;,\n",
       "       &#x27;simulated5.bw&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-3b3aec41-2618-4e74-822c-974a57577bdc' class='xr-section-summary-in' type='checkbox'  checked><label for='section-3b3aec41-2618-4e74-822c-974a57577bdc' class='xr-section-summary' >Data variables: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>cov</span></div><div class='xr-var-dims'>(_sequence, cov_sample)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>dask.array&lt;chunksize=(14, 1), meta=np.ndarray&gt;</div><input id='attrs-8a89d868-88f1-4f42-b753-f95438d61d6d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-8a89d868-88f1-4f42-b753-f95438d61d6d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-a18fbbcf-6a5f-4b19-aa61-fba5c53e07a3' class='xr-var-data-in' type='checkbox'><label for='data-a18fbbcf-6a5f-4b19-aa61-fba5c53e07a3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><table>\n",
       "    <tr>\n",
       "        <td>\n",
       "            <table style=\"border-collapse: collapse;\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <td> </td>\n",
       "                        <th> Array </th>\n",
       "                        <th> Chunk </th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Bytes </th>\n",
       "                        <td> 560 B </td>\n",
       "                        <td> 112 B </td>\n",
       "                    </tr>\n",
       "                    \n",
       "                    <tr>\n",
       "                        <th> Shape </th>\n",
       "                        <td> (14, 5) </td>\n",
       "                        <td> (14, 1) </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Dask graph </th>\n",
       "                        <td colspan=\"2\"> 5 chunks in 2 graph layers </td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th> Data type </th>\n",
       "                        <td colspan=\"2\"> object numpy.ndarray </td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </td>\n",
       "        <td>\n",
       "        <svg width=\"94\" height=\"170\" style=\"stroke:rgb(0,0,0);stroke-width:1\" >\n",
       "\n",
       "  <!-- Horizontal lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"44\" y2=\"0\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"0\" y1=\"120\" x2=\"44\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Vertical lines -->\n",
       "  <line x1=\"0\" y1=\"0\" x2=\"0\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "  <line x1=\"8\" y1=\"0\" x2=\"8\" y2=\"120\" />\n",
       "  <line x1=\"17\" y1=\"0\" x2=\"17\" y2=\"120\" />\n",
       "  <line x1=\"26\" y1=\"0\" x2=\"26\" y2=\"120\" />\n",
       "  <line x1=\"35\" y1=\"0\" x2=\"35\" y2=\"120\" />\n",
       "  <line x1=\"44\" y1=\"0\" x2=\"44\" y2=\"120\" style=\"stroke-width:2\" />\n",
       "\n",
       "  <!-- Colored Rectangle -->\n",
       "  <polygon points=\"0.0,0.0 44.09664557982572,0.0 44.09664557982572,120.0 0.0,120.0\" style=\"fill:#ECB172A0;stroke-width:0\"/>\n",
       "\n",
       "  <!-- Text -->\n",
       "  <text x=\"22.048323\" y=\"140.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" >5</text>\n",
       "  <text x=\"64.096646\" y=\"60.000000\" font-size=\"1.0rem\" font-weight=\"100\" text-anchor=\"middle\" transform=\"rotate(0,64.096646,60.000000)\">14</text>\n",
       "</svg>\n",
       "        </td>\n",
       "    </tr>\n",
       "</table></div></li></ul></div></li><li class='xr-section-item'><input id='section-72f1bcea-273c-445f-8767-14b6944ed4ab' class='xr-section-summary-in' type='checkbox'  ><label for='section-72f1bcea-273c-445f-8767-14b6944ed4ab' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>cov_sample</div></div><div class='xr-index-preview'>PandasIndex</div><div></div><input id='index-4e2b2b5b-e7e7-4690-b3cb-6ebf85b6436d' class='xr-index-data-in' type='checkbox'/><label for='index-4e2b2b5b-e7e7-4690-b3cb-6ebf85b6436d' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;simulated1.bw&#x27;, &#x27;simulated2.bw&#x27;, &#x27;simulated3.bw&#x27;, &#x27;simulated4.bw&#x27;,\n",
       "       &#x27;simulated5.bw&#x27;],\n",
       "      dtype=&#x27;object&#x27;, name=&#x27;cov_sample&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-ba21b0e9-107b-40a5-aeca-307da4091c9f' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-ba21b0e9-107b-40a5-aeca-307da4091c9f' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset> Size: 600B\n",
       "Dimensions:     (_sequence: 14, cov_sample: 5)\n",
       "Coordinates:\n",
       "  * cov_sample  (cov_sample) object 40B 'simulated1.bw' ... 'simulated5.bw'\n",
       "Dimensions without coordinates: _sequence\n",
       "Data variables:\n",
       "    cov         (_sequence, cov_sample) object 560B dask.array<chunksize=(14, 1), meta=np.ndarray>"
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "corrupt buffer, data seem truncated",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[282], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m cov \u001b[38;5;241m=\u001b[39m \u001b[43mdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcov_sample\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msimulated1.bw\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcov\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\n\u001b[1;32m      2\u001b[0m cov\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/dataarray.py:811\u001b[0m, in \u001b[0;36mDataArray.values\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    798\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m    799\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mvalues\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m np\u001b[38;5;241m.\u001b[39mndarray:\n\u001b[1;32m    800\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    801\u001b[0m \u001b[38;5;124;03m    The array's data converted to numpy.ndarray.\u001b[39;00m\n\u001b[1;32m    802\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    809\u001b[0m \u001b[38;5;124;03m    to this array may be reflected in the DataArray as well.\u001b[39;00m\n\u001b[1;32m    810\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 811\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvariable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/variable.py:554\u001b[0m, in \u001b[0;36mVariable.values\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    551\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m    552\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mvalues\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m np\u001b[38;5;241m.\u001b[39mndarray:\n\u001b[1;32m    553\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"The variable's data as a numpy.ndarray\"\"\"\u001b[39;00m\n\u001b[0;32m--> 554\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_as_array_or_item\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_data\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/variable.py:352\u001b[0m, in \u001b[0;36m_as_array_or_item\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    338\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_as_array_or_item\u001b[39m(data):\n\u001b[1;32m    339\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return the given values as a numpy array, or as an individual item if\u001b[39;00m\n\u001b[1;32m    340\u001b[0m \u001b[38;5;124;03m    it's a 0d datetime64 or timedelta64 array.\u001b[39;00m\n\u001b[1;32m    341\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    350\u001b[0m \u001b[38;5;124;03m    TODO: remove this (replace with np.asarray) once these issues are fixed\u001b[39;00m\n\u001b[1;32m    351\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 352\u001b[0m     data \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(data)\n\u001b[1;32m    353\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    354\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m data\u001b[38;5;241m.\u001b[39mdtype\u001b[38;5;241m.\u001b[39mkind \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mM\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/dask/array/core.py:1700\u001b[0m, in \u001b[0;36mArray.__array__\u001b[0;34m(self, dtype, **kwargs)\u001b[0m\n\u001b[1;32m   1699\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__array__\u001b[39m(\u001b[38;5;28mself\u001b[39m, dtype\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m-> 1700\u001b[0m     x \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1701\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m dtype \u001b[38;5;129;01mand\u001b[39;00m x\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m!=\u001b[39m dtype:\n\u001b[1;32m   1702\u001b[0m         x \u001b[38;5;241m=\u001b[39m x\u001b[38;5;241m.\u001b[39mastype(dtype)\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/dask/base.py:375\u001b[0m, in \u001b[0;36mDaskMethodsMixin.compute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    351\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    352\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Compute this dask collection\u001b[39;00m\n\u001b[1;32m    353\u001b[0m \n\u001b[1;32m    354\u001b[0m \u001b[38;5;124;03m    This turns a lazy Dask collection into its in-memory equivalent.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    373\u001b[0m \u001b[38;5;124;03m    dask.compute\u001b[39;00m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 375\u001b[0m     (result,) \u001b[38;5;241m=\u001b[39m \u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtraverse\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    376\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/dask/base.py:661\u001b[0m, in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    658\u001b[0m     postcomputes\u001b[38;5;241m.\u001b[39mappend(x\u001b[38;5;241m.\u001b[39m__dask_postcompute__())\n\u001b[1;32m    660\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m shorten_traceback():\n\u001b[0;32m--> 661\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43mschedule\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdsk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    663\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m repack([f(r, \u001b[38;5;241m*\u001b[39ma) \u001b[38;5;28;01mfor\u001b[39;00m r, (f, a) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(results, postcomputes)])\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/indexing.py:580\u001b[0m, in \u001b[0;36mImplicitToExplicitIndexingAdapter.__array__\u001b[0;34m(self, dtype, copy)\u001b[0m\n\u001b[1;32m    578\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39masarray(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_duck_array(), dtype\u001b[38;5;241m=\u001b[39mdtype, copy\u001b[38;5;241m=\u001b[39mcopy)\n\u001b[1;32m    579\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 580\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m np\u001b[38;5;241m.\u001b[39masarray(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_duck_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m, dtype\u001b[38;5;241m=\u001b[39mdtype)\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/indexing.py:583\u001b[0m, in \u001b[0;36mImplicitToExplicitIndexingAdapter.get_duck_array\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    582\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_duck_array\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 583\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marray\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_duck_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/indexing.py:794\u001b[0m, in \u001b[0;36mCopyOnWriteArray.get_duck_array\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    793\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_duck_array\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 794\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marray\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_duck_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/indexing.py:657\u001b[0m, in \u001b[0;36mLazilyIndexedArray.get_duck_array\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    653\u001b[0m     array \u001b[38;5;241m=\u001b[39m apply_indexer(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39marray, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkey)\n\u001b[1;32m    654\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    655\u001b[0m     \u001b[38;5;66;03m# If the array is not an ExplicitlyIndexedNDArrayMixin,\u001b[39;00m\n\u001b[1;32m    656\u001b[0m     \u001b[38;5;66;03m# it may wrap a BackendArray so use its __getitem__\u001b[39;00m\n\u001b[0;32m--> 657\u001b[0m     array \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marray\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m    659\u001b[0m \u001b[38;5;66;03m# self.array[self.key] is now a numpy array when\u001b[39;00m\n\u001b[1;32m    660\u001b[0m \u001b[38;5;66;03m# self.array is a BackendArray subclass\u001b[39;00m\n\u001b[1;32m    661\u001b[0m \u001b[38;5;66;03m# and self.key is BasicIndexer((slice(None, None, None),))\u001b[39;00m\n\u001b[1;32m    662\u001b[0m \u001b[38;5;66;03m# so we need the explicit check for ExplicitlyIndexed\u001b[39;00m\n\u001b[1;32m    663\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(array, ExplicitlyIndexed):\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/backends/zarr.py:166\u001b[0m, in \u001b[0;36mZarrArrayWrapper.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    164\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, indexing\u001b[38;5;241m.\u001b[39mOuterIndexer):\n\u001b[1;32m    165\u001b[0m     method \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_oindex\n\u001b[0;32m--> 166\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mindexing\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexplicit_indexing_adapter\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    167\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43marray\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexing\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mIndexingSupport\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mVECTORIZED\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\n\u001b[1;32m    168\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/core/indexing.py:1018\u001b[0m, in \u001b[0;36mexplicit_indexing_adapter\u001b[0;34m(key, shape, indexing_support, raw_indexing_method)\u001b[0m\n\u001b[1;32m    996\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Support explicit indexing by delegating to a raw indexing method.\u001b[39;00m\n\u001b[1;32m    997\u001b[0m \n\u001b[1;32m    998\u001b[0m \u001b[38;5;124;03mOuter and/or vectorized indexers are supported by indexing a second time\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1015\u001b[0m \u001b[38;5;124;03mIndexing result, in the form of a duck numpy-array.\u001b[39;00m\n\u001b[1;32m   1016\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1017\u001b[0m raw_key, numpy_indices \u001b[38;5;241m=\u001b[39m decompose_indexer(key, shape, indexing_support)\n\u001b[0;32m-> 1018\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mraw_indexing_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mraw_key\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtuple\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1019\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m numpy_indices\u001b[38;5;241m.\u001b[39mtuple:\n\u001b[1;32m   1020\u001b[0m     \u001b[38;5;66;03m# index the loaded np.ndarray\u001b[39;00m\n\u001b[1;32m   1021\u001b[0m     indexable \u001b[38;5;241m=\u001b[39m NumpyIndexingAdapter(result)\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/xarray/backends/zarr.py:156\u001b[0m, in \u001b[0;36mZarrArrayWrapper._getitem\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    155\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_getitem\u001b[39m(\u001b[38;5;28mself\u001b[39m, key):\n\u001b[0;32m--> 156\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_array\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/zarr/core.py:795\u001b[0m, in \u001b[0;36mArray.__getitem__\u001b[0;34m(self, selection)\u001b[0m\n\u001b[1;32m    793\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mvindex[selection]\n\u001b[1;32m    794\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_pure_orthogonal_indexing(pure_selection, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mndim):\n\u001b[0;32m--> 795\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_orthogonal_selection\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpure_selection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfields\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    797\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_basic_selection(pure_selection, fields\u001b[38;5;241m=\u001b[39mfields)\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/zarr/core.py:1077\u001b[0m, in \u001b[0;36mArray.get_orthogonal_selection\u001b[0;34m(self, selection, out, fields)\u001b[0m\n\u001b[1;32m   1074\u001b[0m \u001b[38;5;66;03m# setup indexer\u001b[39;00m\n\u001b[1;32m   1075\u001b[0m indexer \u001b[38;5;241m=\u001b[39m OrthogonalIndexer(selection, \u001b[38;5;28mself\u001b[39m)\n\u001b[0;32m-> 1077\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_selection\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexer\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfields\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/zarr/core.py:1340\u001b[0m, in \u001b[0;36mArray._get_selection\u001b[0;34m(self, indexer, out, fields)\u001b[0m\n\u001b[1;32m   1337\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m math\u001b[38;5;241m.\u001b[39mprod(out_shape) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1338\u001b[0m     \u001b[38;5;66;03m# allow storage to get multiple items at once\u001b[39;00m\n\u001b[1;32m   1339\u001b[0m     lchunk_coords, lchunk_selection, lout_selection \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;241m*\u001b[39mindexer)\n\u001b[0;32m-> 1340\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_chunk_getitems\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1341\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlchunk_coords\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1342\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlchunk_selection\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1343\u001b[0m \u001b[43m        \u001b[49m\u001b[43mout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1344\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlout_selection\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1345\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdrop_axes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindexer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop_axes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1346\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1347\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1348\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m out\u001b[38;5;241m.\u001b[39mshape:\n\u001b[1;32m   1349\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m out\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/zarr/core.py:2185\u001b[0m, in \u001b[0;36mArray._chunk_getitems\u001b[0;34m(self, lchunk_coords, lchunk_selection, out, lout_selection, drop_axes, fields)\u001b[0m\n\u001b[1;32m   2183\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m ckey, chunk_select, out_select \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(ckeys, lchunk_selection, lout_selection):\n\u001b[1;32m   2184\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ckey \u001b[38;5;129;01min\u001b[39;00m cdatas:\n\u001b[0;32m-> 2185\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_process_chunk\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2186\u001b[0m \u001b[43m            \u001b[49m\u001b[43mout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2187\u001b[0m \u001b[43m            \u001b[49m\u001b[43mcdatas\u001b[49m\u001b[43m[\u001b[49m\u001b[43mckey\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2188\u001b[0m \u001b[43m            \u001b[49m\u001b[43mchunk_select\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2189\u001b[0m \u001b[43m            \u001b[49m\u001b[43mdrop_axes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2190\u001b[0m \u001b[43m            \u001b[49m\u001b[43mout_is_ndarray\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2191\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2192\u001b[0m \u001b[43m            \u001b[49m\u001b[43mout_select\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2193\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpartial_read_decode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpartial_read_decode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2194\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2195\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   2196\u001b[0m         \u001b[38;5;66;03m# check exception type\u001b[39;00m\n\u001b[1;32m   2197\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fill_value \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/zarr/core.py:2098\u001b[0m, in \u001b[0;36mArray._process_chunk\u001b[0;34m(self, out, cdata, chunk_selection, drop_axes, out_is_ndarray, fields, out_selection, partial_read_decode)\u001b[0m\n\u001b[1;32m   2096\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ArrayIndexError:\n\u001b[1;32m   2097\u001b[0m     cdata \u001b[38;5;241m=\u001b[39m cdata\u001b[38;5;241m.\u001b[39mread_full()\n\u001b[0;32m-> 2098\u001b[0m chunk \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_decode_chunk\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2100\u001b[0m \u001b[38;5;66;03m# select data from chunk\u001b[39;00m\n\u001b[1;32m   2101\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m fields:\n",
      "File \u001b[0;32m~/opt/miniconda3/envs/eugene_tools/lib/python3.11/site-packages/zarr/core.py:2361\u001b[0m, in \u001b[0;36mArray._decode_chunk\u001b[0;34m(self, cdata, start, nitems, expected_shape)\u001b[0m\n\u001b[1;32m   2359\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_filters:\n\u001b[1;32m   2360\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m f \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mreversed\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_filters):\n\u001b[0;32m-> 2361\u001b[0m         chunk \u001b[38;5;241m=\u001b[39m \u001b[43mf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdecode\u001b[49m\u001b[43m(\u001b[49m\u001b[43mchunk\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2363\u001b[0m \u001b[38;5;66;03m# view as numpy array with correct dtype\u001b[39;00m\n\u001b[1;32m   2364\u001b[0m chunk \u001b[38;5;241m=\u001b[39m ensure_ndarray_like(chunk)\n",
      "File \u001b[0;32mnumcodecs/vlen.pyx:449\u001b[0m, in \u001b[0;36mnumcodecs.vlen.VLenArray.decode\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: corrupt buffer, data seem truncated"
     ]
    }
   ],
   "source": [
    "cov = data.sel(cov_sample=\"simulated1.bw\").cov.values\n",
    "cov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 26630.50it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Region chr1:4-24: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1]\n",
      "Region chr1:47-67: [3.0, 3.0, 3.0, 2.0, 2.0, 1.0, 2.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 1.0, 1.0, 1.0, 1.0], [3 3 3 2 2 1 2 1 1 1 1 1 2 2 2 2 1 1 1 1]\n",
      "Region chr2:46-66: [2.0, 1.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 2.0, 4.0, 4.0, 3.0, 3.0, 5.0, 5.0], [2 1 1 2 2 2 2 2 2 2 2 3 3 2 4 4 3 3 5 5]\n",
      "Region chr2:174-194: [2.0, 1.0, 1.0, 1.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 5.0, 6.0, 6.0, 6.0, 5.0, 4.0, 3.0, 2.0, 3.0, 4.0], [2 1 1 1 2 3 3 4 4 4 5 6 6 6 5 4 3 2 3 4]\n",
      "Region chr3:18-38: [1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0], [1 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 0 0 0 0]\n",
      "Region chr3:78-98: [1.0, 0.0, 0.0, 1.0, 3.0, 4.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 5.0, 4.0, 2.0, 1.0, 0.0, 0.0, 1.0, 3.0], [1 0 0 1 3 4 5 5 5 5 5 5 5 4 2 1 0 0 1 3]\n",
      "Region chr4:35-55: [1.0, 0.0, 0.0, 0.0, 2.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 4.0, 2.0, 2.0, 2.0, 2.0, 1.0, 4.0], [1 0 0 0 2 2 3 3 4 4 4 4 4 4 2 2 2 2 1 4]\n",
      "Region chr4:87-107: [2.0, 3.0, 4.0, 4.0, 4.0, 4.0, 3.0, 2.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 7.0, 6.0, 6.0], [2 3 4 4 4 4 3 2 3 3 3 4 4 4 4 4 5 7 6 6]\n",
      "Region chr5:40-60: [1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 2.0, 1.0, 1.0], [1 1 0 0 0 0 0 1 2 2 2 2 2 2 2 3 3 2 1 1]\n",
      "Region chr5:156-176: [1.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 2.0, 2.0, 4.0, 4.0, 4.0, 6.0, 6.0, 6.0, 6.0], [1 2 2 2 2 2 2 2 2 2 3 2 2 4 4 4 6 6 6 6]\n",
      "Region chr6:19-39: [3.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 1.0, 2.0, 2.0, 3.0, 3.0, 5.0, 4.0, 4.0, 4.0, 4.0], [3 2 2 2 2 2 2 2 2 2 1 2 2 3 3 5 4 4 4 4]\n",
      "Region chr6:61-81: [6.0, 6.0, 5.0, 5.0, 5.0, 5.0, 4.0, 5.0, 5.0, 6.0, 4.0, 4.0, 4.0, 4.0, 3.0, 4.0, 4.0, 3.0, 3.0, 3.0], [6 6 5 5 5 5 4 5 5 6 4 4 4 4 3 4 4 3 3 3]\n",
      "Region chr7:12-32: [0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 2.0, 2.0, 1.0, 2.0, 2.0, 2.0, 2.0, 3.0], [0 0 0 0 1 1 1 1 1 1 1 1 2 2 1 2 2 2 2 3]\n",
      "Region chr7:153-173: [2.0, 2.0, 2.0, 2.0, 2.0, 3.0, 3.0, 4.0, 4.0, 4.0, 4.0, 4.0, 5.0, 5.0, 6.0, 5.0, 5.0, 5.0, 5.0, 6.0], [2 2 2 2 2 3 3 4 4 4 4 4 5 5 6 5 5 5 5 6]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Make chromsizes a list of tuples\n",
    "chromsizes = {}\n",
    "with open(Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'variable.chrom.sizes') as f:\n",
    "    for line in f:\n",
    "        chrom, size = line.strip().split()\n",
    "        chromsizes[chrom] = int(size)\n",
    "chromsizes\n",
    "\n",
    "bigwig_reader = BigWig(\n",
    "    name=\"cov\",\n",
    "    bigwigs=bigwigs,\n",
    "    samples=[bigwig.name for bigwig in bigwigs],\n",
    "    batch_size=50\n",
    ")\n",
    "assert isinstance(bigwig_reader, BigWig), \"bigwig reader instantiation failed.\"\n",
    "\n",
    "iterator = bigwig_reader._reader(\n",
    "    bed=bed,\n",
    "    f=pyBigWig.open(str(bigwigs[0])),\n",
    "    contig_lengths=chromsizes\n",
    ")\n",
    "\n",
    "_read = [list(seq) for seq in iterator]\n",
    "\n",
    "# Verify that the data matches the expected coverage\n",
    "for i, (region, coverage) in enumerate(new_cov[bigwigs[0].name.replace(\".bw\", \".bam\")].items()):\n",
    "    print(f\"Region {region}: {list(_read[i])}, {coverage}\")\n",
    "    assert np.array_equal(coverage, _read[i]), f\"Region {region} in {bigwigs[0].name} does not match.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `from_region_files`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "from seqdata import from_region_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_from_region_files(\n",
    "    *readers, \n",
    "    names,\n",
    "    true_seqs,\n",
    "    true_covs,\n",
    "    temp_dir, \n",
    "    fixed_length=False, \n",
    "    length_dim=None\n",
    "):\n",
    "    \"\"\"\n",
    "    Tests the FlatFASTA class for a single input FASTA file.\n",
    "\n",
    "    Parameters:\n",
    "        input_fasta (str or Path): Path to the input FASTA file.\n",
    "        temp_dir (str or Path): Directory where temporary files will be stored.\n",
    "        batch_size (int): Batch size for the FlatFASTA reader.\n",
    "        fixed_length (bool): Whether to use fixed-length sequences when writing to Zarr.\n",
    "        length_dim (str): Name of the length dimension in the Zarr output.\n",
    "\n",
    "    Raises:\n",
    "        AssertionError: If any test fails.\n",
    "    \"\"\"\n",
    "    temp_dir = Path(temp_dir)\n",
    "    temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "    out = temp_dir / 'output.zarr'\n",
    "\n",
    "    data = from_region_files(\n",
    "        *readers,\n",
    "        path=out,\n",
    "        sequence_dim=\"_sequence\",\n",
    "        fixed_length=fixed_length,\n",
    "        overwrite=True,\n",
    "        length_dim=length_dim\n",
    "    )\n",
    "\n",
    "    # Verify that the data is the same\n",
    "    for i, name in enumerate(names):\n",
    "        if fixed_length:\n",
    "            seqs = [''.join(row.astype(str)) for row in data[name].values]\n",
    "        else:\n",
    "            seqs = data[name].values.astype(str)\n",
    "        print(name)\n",
    "        print(seqs)\n",
    "        print(true[i])\n",
    "        assert np.array_equal(true[i], seqs), \"Sequences do not match.\"\n",
    "\n",
    "    # Clean up temporary files\n",
    "    for f in temp_dir.iterdir():\n",
    "        print(f)\n",
    "        if f.is_dir():\n",
    "            os.system(f'rm -r {f}')\n",
    "        else:\n",
    "            f.unlink()\n",
    "\n",
    "    print(\"All tests passed successfully!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixed_length = 20\n",
    "\n",
    "outfile = Path(sd.__file__).resolve().parent.parent / 'tests' / 'data' / 'tmp' / 'output.zarr'\n",
    "\n",
    "bed = pd.read_csv(fixed_bed_in, sep=\"\\t\", header=None)\n",
    "bed\n",
    "\n",
    "if fixed_length:\n",
    "    midpoint = (bed[1] + bed[2]) // 2\n",
    "    bed[1] = midpoint - fixed_length // 2\n",
    "    bed[2] = midpoint + fixed_length // 2\n",
    "bed\n",
    "\n",
    "fasta_sequences = read_fasta(variable_fasta_in)\n",
    "\n",
    "true_seqs = [fasta_sequences[chrom][start:end] for chrom, start, end in bed.values]\n",
    "true_seqs\n",
    "\n",
    "genome_fasta = GenomeFASTA(\n",
    "    name=\"seq\",\n",
    "    fasta=variable_fasta_in,\n",
    "    batch_size=50,\n",
    ")\n",
    "bam = BAM(\n",
    "    name=\"cov\",\n",
    "    bams=bams,\n",
    "    samples=[bam.name for bam in bams],\n",
    "    batch_size=50,\n",
    ")\n",
    "\n",
    "data = from_region_files(\n",
    "    genome_fasta, \n",
    "    bam, \n",
    "    bed=fixed_bed_in,\n",
    "    path=outfile,\n",
    "    fixed_length=20,\n",
    "    length_dim=\"_length\",\n",
    "    sequence_dim=\"_sequence\",\n",
    "    overwrite=True\n",
    ")\n",
    "\n",
    "# Verify length and sequence dimensions\n",
    "if fixed_length:\n",
    "    assert data[\"seq\"].shape[1] == fixed_length, \"Length dimension is incorrect.\"\n",
    "assert data[\"seq\"].shape[0] == len(bed), \"Sequence dimension is incorrect.\"\n",
    "\n",
    "# Verify that the data is the same\n",
    "if fixed_length:\n",
    "    seqs = [''.join(row.astype(str)) for row in data[\"seq\"].values]\n",
    "else:\n",
    "    seqs = data[\"seq\"].values.astype(str)\n",
    "print(seqs)\n",
    "print(true_seqs)\n",
    "assert np.array_equal(true_seqs, seqs), \"Sequences do not match.\"\n",
    "\n",
    "# Verify that the data matches the expected coverage for all regions\n",
    "for file_index, bam_file in enumerate(bams):\n",
    "    print(f\"Testing {bam_file}\")\n",
    "    bam_name = bam_file.name\n",
    "    for region_index, (region, coverage) in enumerate(new_cov[bam_name].items()):\n",
    "        print(data['cov'][region_index].shape)\n",
    "        print(f\"Region {region}: {list(data['cov'].values[region_index, file_index])}, {coverage}\")\n",
    "        assert np.array_equal(coverage, data[\"cov\"].values[region_index, file_index]), f\"Region {region} in {bam_name} does not match.\"\n",
    "\n",
    "# Clean up temporary files\n",
    "for f in temp_dir.iterdir():\n",
    "    if f.is_dir():\n",
    "        os.system(f'rm -r {f}')\n",
    "    else:\n",
    "        f.unlink()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `read_genome_fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "from seqdata import read_genome_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 82472.27it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "temp_dir = Path(out)\n",
    "temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "out = temp_dir / 'output.zarr'\n",
    "\n",
    "sdata = read_genome_fasta(\n",
    "    fasta=variable_fasta_in,\n",
    "    bed=fixed_bed_in,\n",
    "    out=out,\n",
    "    name=\"seq\",\n",
    "    fixed_length=20,\n",
    "    batch_size=50,\n",
    "    overwrite=True,\n",
    ")\n",
    "\n",
    "# Verify length and sequence dimensions\n",
    "fasta_sequences = read_fasta(variable_fasta_in)\n",
    "true = [fasta_sequences[chrom][start:end] for chrom, start, end in bed.values]\n",
    "\n",
    "if fixed_length:\n",
    "    assert data[\"seq\"].shape[1] == fixed_length, \"Length dimension is incorrect.\"\n",
    "assert data[\"seq\"].shape[0] == len(bed), \"Sequence dimension is incorrect.\"\n",
    "\n",
    "# Verify that the data is the same\n",
    "if fixed_length:\n",
    "    seqs = [''.join(row.astype(str)) for row in data[\"seq\"].values]\n",
    "else:\n",
    "    seqs = data[\"seq\"].values.astype(str)\n",
    "assert np.array_equal(true, seqs), \"Sequences do not match.\"\n",
    "\n",
    "# Clean up temporary files\n",
    "for f in temp_dir.iterdir():\n",
    "    if f.is_dir():\n",
    "        os.system(f'rm -r {f}')\n",
    "    else:\n",
    "        f.unlink()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `read_bam`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "from seqdata import read_bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 123361.88it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 14135.83it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 11027.28it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 11056.35it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 13862.19it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 12959.67it/s]\n"
     ]
    }
   ],
   "source": [
    "temp_dir = Path(out)\n",
    "temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "out = temp_dir / 'output.zarr'\n",
    "\n",
    "sdata = read_bam(\n",
    "    fasta=variable_fasta_in,\n",
    "    bed=fixed_bed_in,\n",
    "    bams=bams,\n",
    "    out=out,\n",
    "    seq_name=\"seq\",\n",
    "    cov_name=\"cov\",\n",
    "    samples=[b.name for b in bams],\n",
    "    fixed_length=20,\n",
    "    batch_size=50,\n",
    "    overwrite=True,\n",
    ")\n",
    "\n",
    "# Verify length and sequence dimensions\n",
    "fasta_sequences = read_fasta(variable_fasta_in)\n",
    "true = [fasta_sequences[chrom][start:end] for chrom, start, end in bed.values]\n",
    "\n",
    "if fixed_length:\n",
    "    assert data[\"seq\"].shape[1] == fixed_length, \"Length dimension is incorrect.\"\n",
    "assert data[\"seq\"].shape[0] == len(bed), \"Sequence dimension is incorrect.\"\n",
    "\n",
    "# Verify that the data is the same\n",
    "if fixed_length:\n",
    "    seqs = [''.join(row.astype(str)) for row in data[\"seq\"].values]\n",
    "else:\n",
    "    seqs = data[\"seq\"].values.astype(str)\n",
    "assert np.array_equal(true, seqs), \"Sequences do not match.\"\n",
    "\n",
    "# Clean up temporary files\n",
    "for f in temp_dir.iterdir():\n",
    "    if f.is_dir():\n",
    "        os.system(f'rm -r {f}')\n",
    "    else:\n",
    "        f.unlink()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `read_bigwig`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "from seqdata import read_bigwig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['simulated1.bw',\n",
       " 'simulated2.bw',\n",
       " 'simulated3.bw',\n",
       " 'simulated4.bw',\n",
       " 'simulated5.bw']"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[b.name for b in bigwigs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 136877.05it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 16105.39it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 29360.13it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 24734.73it/s]\n",
      "100%|██████████| 14/14 [00:00<00:00, 38990.87it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 14/14 [00:00<00:00, 26985.41it/s]\n"
     ]
    }
   ],
   "source": [
    "temp_dir = Path(out)\n",
    "temp_dir.mkdir(parents=True, exist_ok=True)\n",
    "out = temp_dir / 'output.zarr'\n",
    "\n",
    "sdata = read_bigwig(\n",
    "    fasta=variable_fasta_in,\n",
    "    bed=fixed_bed_in,\n",
    "    bigwigs=bigwigs,\n",
    "    out=out,\n",
    "    seq_name=\"seq\",\n",
    "    cov_name=\"cov\",\n",
    "    samples=[b.name for b in bigwigs],\n",
    "    fixed_length=20,\n",
    "    batch_size=50,\n",
    "    overwrite=True,\n",
    ")\n",
    "\n",
    "# Verify length and sequence dimensions\n",
    "fasta_sequences = read_fasta(variable_fasta_in)\n",
    "true = [fasta_sequences[chrom][start:end] for chrom, start, end in bed.values]\n",
    "\n",
    "if fixed_length:\n",
    "    assert data[\"seq\"].shape[1] == fixed_length, \"Length dimension is incorrect.\"\n",
    "assert data[\"seq\"].shape[0] == len(bed), \"Sequence dimension is incorrect.\"\n",
    "\n",
    "# Verify that the data is the same\n",
    "if fixed_length:\n",
    "    seqs = [''.join(row.astype(str)) for row in data[\"seq\"].values]\n",
    "else:\n",
    "    seqs = data[\"seq\"].values.astype(str)\n",
    "assert np.array_equal(true, seqs), \"Sequences do not match.\"\n",
    "\n",
    "# Clean up temporary files\n",
    "for f in temp_dir.iterdir():\n",
    "    if f.is_dir():\n",
    "        os.system(f'rm -r {f}')\n",
    "    else:\n",
    "        f.unlink()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DONE!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing BAM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrom: chr1, start: 10, end: 30\n",
      "20\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([2, 2, 2, 2, 3, 4, 4, 4, 4, 3, 3, 3, 4, 6, 6, 6, 6, 6, 6, 7],\n",
       "      dtype=uint16)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "seq_num = 0\n",
    "print(f\"chrom: {bed.iloc[seq_num]['chrom']}, start: {bed.iloc[seq_num]['start']}, end: {bed.iloc[seq_num]['end']}\")\n",
    "cov = bam_reader._count_depth_only(\n",
    "    f=pysam.AlignmentFile(inbam),\n",
    "    contig=bed.iloc[seq_num]['chrom'],\n",
    "    start=bed.iloc[seq_num]['start'],\n",
    "    end=bed.iloc[seq_num]['end']\n",
    ")\n",
    "print(len(cov))\n",
    "cov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "_read = [seq for seq in iterator]\n",
    "_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#inbam = \"/cellar/users/aklie/data/datasets/SeqDatasets/K562_ATAC-seq/data/merged.bam\"\n",
    "#inbed = \"/cellar/users/aklie/data/datasets/SeqDatasets/K562_ATAC-seq/data/ENCSR868FGK_K562_ATAC-seq_peaks.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = pysam.AlignmentFile(inbam)\n",
    "bed = pd.read_csv(inbed, sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array('L', [0, 3, 0, 0, 0, 0, 0, 4, 0, 0, 4, 0, 0, 6, 0, 0, 6, 0, 0, 0]),\n",
       " array('L', [0, 0, 2, 0, 2, 2, 0, 0, 4, 5, 0, 0, 0, 0, 6, 0, 0, 0, 6, 6]),\n",
       " array('L', [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0, 0, 0, 0, 0, 0, 0]),\n",
       " array('L', [4, 0, 0, 2, 0, 0, 3, 0, 0, 0, 0, 4, 0, 0, 0, 7, 0, 5, 0, 0]))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "f.count_coverage(\n",
    "    contig=bed.iloc[1][0],\n",
    "    start=bed.iloc[1][1],\n",
    "    stop=bed.iloc[1][2],\n",
    "    #read_callback='nofilter',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the data using the reader\n",
    "_read = [seq.decode('utf-8') for seq in iterator]\n",
    "\n",
    "# Verify that the data is the same\n",
    "assert np.array_equal(_read, true), \"GenomeFASTA reader failed to read in the correct values.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read name: read_chr1_9_19\n",
      "Reference: chr1\n",
      "Start position: 9\n",
      "Read sequence: CCGACTAACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_10_20\n",
      "Reference: chr1\n",
      "Start position: 10\n",
      "Read sequence: CGACTAACTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_14_24\n",
      "Reference: chr1\n",
      "Start position: 14\n",
      "Read sequence: TAACTGACTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_15_25\n",
      "Reference: chr1\n",
      "Start position: 15\n",
      "Read sequence: AACTGACTGA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_20_30\n",
      "Reference: chr1\n",
      "Start position: 20\n",
      "Read sequence: ACTGATGATG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_22_32\n",
      "Reference: chr1\n",
      "Start position: 22\n",
      "Read sequence: TGATGATGAT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_23_33\n",
      "Reference: chr1\n",
      "Start position: 23\n",
      "Read sequence: GATGATGATG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_23_33\n",
      "Reference: chr1\n",
      "Start position: 23\n",
      "Read sequence: GATGATGATG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_9_19\n",
      "Reference: chr1\n",
      "Start position: 24\n",
      "Read sequence: ATGATGATGC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_10_20\n",
      "Reference: chr1\n",
      "Start position: 25\n",
      "Read sequence: TGATGATGCA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_14_24\n",
      "Reference: chr1\n",
      "Start position: 29\n",
      "Read sequence: GATGCATGCT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_15_25\n",
      "Reference: chr1\n",
      "Start position: 30\n",
      "Read sequence: ATGCATGCTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_20_30\n",
      "Reference: chr1\n",
      "Start position: 35\n",
      "Read sequence: TGCTGATGCT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_22_32\n",
      "Reference: chr1\n",
      "Start position: 37\n",
      "Read sequence: CTGATGCTGA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_23_33\n",
      "Reference: chr1\n",
      "Start position: 38\n",
      "Read sequence: TGATGCTGAA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_23_33\n",
      "Reference: chr1\n",
      "Start position: 38\n",
      "Read sequence: TGATGCTGAA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_71_81\n",
      "Reference: chr1\n",
      "Start position: 71\n",
      "Read sequence: GACTGACTGT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_72_82\n",
      "Reference: chr1\n",
      "Start position: 72\n",
      "Read sequence: ACTGACTGTA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_74_84\n",
      "Reference: chr1\n",
      "Start position: 74\n",
      "Read sequence: TGACTGTACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_80_90\n",
      "Reference: chr1\n",
      "Start position: 80\n",
      "Read sequence: TACTCCTACC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_84_94\n",
      "Reference: chr1\n",
      "Start position: 84\n",
      "Read sequence: CCTACCATGA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_71_81\n",
      "Reference: chr1\n",
      "Start position: 86\n",
      "Read sequence: TACCATGACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_72_82\n",
      "Reference: chr1\n",
      "Start position: 87\n",
      "Read sequence: ACCATGACTA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_74_84\n",
      "Reference: chr1\n",
      "Start position: 89\n",
      "Read sequence: CATGACTATC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_92_102\n",
      "Reference: chr1\n",
      "Start position: 92\n",
      "Read sequence: GACTATCCTA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_93_103\n",
      "Reference: chr1\n",
      "Start position: 93\n",
      "Read sequence: ACTATCCTAG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_94_104\n",
      "Reference: chr1\n",
      "Start position: 94\n",
      "Read sequence: CTATCCTAGT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_80_90\n",
      "Reference: chr1\n",
      "Start position: 95\n",
      "Read sequence: TATCCTAGTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_98_108\n",
      "Reference: chr1\n",
      "Start position: 98\n",
      "Read sequence: CCTAGTGCTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_84_94\n",
      "Reference: chr1\n",
      "Start position: 99\n",
      "Read sequence: CTAGTGCTGA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_92_102\n",
      "Reference: chr1\n",
      "Start position: 107\n",
      "Read sequence: GACCTGACTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_93_103\n",
      "Reference: chr1\n",
      "Start position: 108\n",
      "Read sequence: ACCTGACTGA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_94_104\n",
      "Reference: chr1\n",
      "Start position: 109\n",
      "Read sequence: CCTGACTGAT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_98_108\n",
      "Reference: chr1\n",
      "Start position: 113\n",
      "Read sequence: ACTGATGCTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_134_144\n",
      "Reference: chr1\n",
      "Start position: 134\n",
      "Read sequence: ATGCACTGAC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_135_145\n",
      "Reference: chr1\n",
      "Start position: 135\n",
      "Read sequence: TGCACTGACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_135_145\n",
      "Reference: chr1\n",
      "Start position: 135\n",
      "Read sequence: TGCACTGACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_147_157\n",
      "Reference: chr1\n",
      "Start position: 147\n",
      "Read sequence: CTCTACATGA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_148_158\n",
      "Reference: chr1\n",
      "Start position: 148\n",
      "Read sequence: TCTACATGAC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_134_144\n",
      "Reference: chr1\n",
      "Start position: 149\n",
      "Read sequence: CTACATGACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_135_145\n",
      "Reference: chr1\n",
      "Start position: 150\n",
      "Read sequence: TACATGACTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_135_145\n",
      "Reference: chr1\n",
      "Start position: 150\n",
      "Read sequence: TACATGACTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_153_163\n",
      "Reference: chr1\n",
      "Start position: 153\n",
      "Read sequence: ATGACTGACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_154_164\n",
      "Reference: chr1\n",
      "Start position: 154\n",
      "Read sequence: TGACTGACTC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_155_165\n",
      "Reference: chr1\n",
      "Start position: 155\n",
      "Read sequence: GACTGACTCA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_155_165\n",
      "Reference: chr1\n",
      "Start position: 155\n",
      "Read sequence: GACTGACTCA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_159_169\n",
      "Reference: chr1\n",
      "Start position: 159\n",
      "Read sequence: GACTCACTCA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_147_157\n",
      "Reference: chr1\n",
      "Start position: 162\n",
      "Read sequence: TCACTCATCT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_148_158\n",
      "Reference: chr1\n",
      "Start position: 163\n",
      "Read sequence: CACTCATCTG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_153_163\n",
      "Reference: chr1\n",
      "Start position: 168\n",
      "Read sequence: ATCTGACATA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_154_164\n",
      "Reference: chr1\n",
      "Start position: 169\n",
      "Read sequence: TCTGACATAT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_155_165\n",
      "Reference: chr1\n",
      "Start position: 170\n",
      "Read sequence: CTGACATATC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_155_165\n",
      "Reference: chr1\n",
      "Start position: 170\n",
      "Read sequence: CTGACATATC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr1_159_169\n",
      "Reference: chr1\n",
      "Start position: 174\n",
      "Read sequence: CATATCCATG\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_30_40\n",
      "Reference: chr2\n",
      "Start position: 30\n",
      "Read sequence: ATCTACTACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_34_44\n",
      "Reference: chr2\n",
      "Start position: 34\n",
      "Read sequence: ACTACTGCTA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_42_52\n",
      "Reference: chr2\n",
      "Start position: 42\n",
      "Read sequence: TATACTCATA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_42_52\n",
      "Reference: chr2\n",
      "Start position: 42\n",
      "Read sequence: TATACTCATA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_44_54\n",
      "Reference: chr2\n",
      "Start position: 44\n",
      "Read sequence: TACTCATATC\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_30_40\n",
      "Reference: chr2\n",
      "Start position: 45\n",
      "Read sequence: ACTCATATCT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_34_44\n",
      "Reference: chr2\n",
      "Start position: 49\n",
      "Read sequence: ATATCTACTA\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_42_52\n",
      "Reference: chr2\n",
      "Start position: 57\n",
      "Read sequence: TACTACTACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_42_52\n",
      "Reference: chr2\n",
      "Start position: 57\n",
      "Read sequence: TACTACTACT\n",
      "CIGAR string: 10M\n",
      "------\n",
      "Read name: read_chr2_44_54\n",
      "Reference: chr2\n",
      "Start position: 59\n",
      "Read sequence: CTACTACTCA\n",
      "CIGAR string: 10M\n",
      "------\n"
     ]
    }
   ],
   "source": [
    "# Open the BAM file for reading\n",
    "with pysam.AlignmentFile(inbam, \"rb\") as bamfile:\n",
    "    # Iterate over each read\n",
    "    for read in bamfile.fetch():\n",
    "        print(f\"Read name: {read.query_name}\")\n",
    "        print(f\"Reference: {bamfile.get_reference_name(read.reference_id)}\")\n",
    "        print(f\"Start position: {read.reference_start}\")\n",
    "        print(f\"Read sequence: {read.query_sequence}\")\n",
    "        print(f\"CIGAR string: {read.cigarstring}\")\n",
    "        print(\"------\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11 eugene_tools",
   "language": "python",
   "name": "eugene_tools"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
